{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Gj7xA44e9bvA"
      },
      "source": [
        "##Intro to Programming Project: Protein TAG Variant Screen"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "cellView": "form",
        "id": "iccGdbe_Pmt9",
        "pycharm": {
          "name": "#%%\n"
        }
      },
      "outputs": [],
      "source": [
        "#@title Install dependencies\n",
        "%%bash -s $use_amber $use_templates\n",
        "\n",
        "set -e\n",
        "\n",
        "USE_AMBER=$1\n",
        "USE_TEMPLATES=$2\n",
        "\n",
        "if [ ! -f COLABFOLD_READY ]; then\n",
        "  # install dependencies\n",
        "  # We have to use \"--no-warn-conflicts\" because colab already has a lot preinstalled with requirements different to ours\n",
        "  pip install -q --no-warn-conflicts \"colabfold[alphafold-minus-jax] @ git+https://github.com/sokrypton/ColabFold\"\n",
        "  # high risk high gain\n",
        "  pip install -q \"jax[cuda11_cudnn805]>=0.3.8,<0.4\" -f https://storage.googleapis.com/jax-releases/jax_releases.html\n",
        "  touch COLABFOLD_READY\n",
        "fi\n",
        "\n",
        "# setup conda\n",
        "if [ ${USE_AMBER} == \"True\" ] || [ ${USE_TEMPLATES} == \"True\" ]; then\n",
        "  if [ ! -f CONDA_READY ]; then\n",
        "    wget -qnc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
        "    bash Miniconda3-latest-Linux-x86_64.sh -bfp /usr/local 2>&1 1>/dev/null\n",
        "    rm Miniconda3-latest-Linux-x86_64.sh\n",
        "    touch CONDA_READY\n",
        "  fi\n",
        "fi\n",
        "# setup template search\n",
        "if [ ${USE_TEMPLATES} == \"True\" ] && [ ! -f HH_READY ]; then\n",
        "  conda install -y -q -c conda-forge -c bioconda kalign2=2.04 hhsuite=3.3.0 python=3.7 2>&1 1>/dev/null\n",
        "  touch HH_READY\n",
        "fi\n",
        "# setup openmm for amber refinement\n",
        "if [ ${USE_AMBER} == \"True\" ] && [ ! -f AMBER_READY ]; then\n",
        "  conda install -y -q -c conda-forge openmm=7.5.1 python=3.7 pdbfixer 2>&1 1>/dev/null\n",
        "  touch AMBER_READY\n",
        "fi"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yE0sYroghJRg",
        "outputId": "fabee71f-4ec9-4c88-f2ff-be9bdc88f91a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: gradio in /usr/local/lib/python3.7/dist-packages (3.3.1)\n",
            "Requirement already satisfied: httpx in /usr/local/lib/python3.7/dist-packages (from gradio) (0.23.0)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.7/dist-packages (from gradio) (6.0)\n",
            "Requirement already satisfied: pydub in /usr/local/lib/python3.7/dist-packages (from gradio) (0.25.1)\n",
            "Requirement already satisfied: python-multipart in /usr/local/lib/python3.7/dist-packages (from gradio) (0.0.5)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.7/dist-packages (from gradio) (3.1.3)\n",
            "Requirement already satisfied: markdown-it-py[linkify,plugins] in /usr/local/lib/python3.7/dist-packages (from gradio) (2.1.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from gradio) (2.28.1)\n",
            "Requirement already satisfied: websockets in /usr/local/lib/python3.7/dist-packages (from gradio) (10.3)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.7/dist-packages (from gradio) (2.11.3)\n",
            "Requirement already satisfied: pycryptodome in /usr/local/lib/python3.7/dist-packages (from gradio) (3.15.0)\n",
            "Requirement already satisfied: orjson in /usr/local/lib/python3.7/dist-packages (from gradio) (3.8.0)\n",
            "Requirement already satisfied: uvicorn in /usr/local/lib/python3.7/dist-packages (from gradio) (0.18.3)\n",
            "Requirement already satisfied: ffmpy in /usr/local/lib/python3.7/dist-packages (from gradio) (0.3.0)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.7/dist-packages (from gradio) (2022.8.2)\n",
            "Requirement already satisfied: h11<0.13,>=0.11 in /usr/local/lib/python3.7/dist-packages (from gradio) (0.12.0)\n",
            "Requirement already satisfied: aiohttp in /usr/local/lib/python3.7/dist-packages (from gradio) (3.8.1)\n",
            "Requirement already satisfied: analytics-python in /usr/local/lib/python3.7/dist-packages (from gradio) (1.4.0)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from gradio) (1.3.5)\n",
            "Requirement already satisfied: pillow in /usr/local/lib/python3.7/dist-packages (from gradio) (7.1.2)\n",
            "Requirement already satisfied: pydantic in /usr/local/lib/python3.7/dist-packages (from gradio) (1.9.2)\n",
            "Requirement already satisfied: fastapi in /usr/local/lib/python3.7/dist-packages (from gradio) (0.85.0)\n",
            "Requirement already satisfied: paramiko in /usr/local/lib/python3.7/dist-packages (from gradio) (2.11.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from gradio) (1.21.6)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (1.2.0)\n",
            "Requirement already satisfied: typing-extensions>=3.7.4 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (4.1.1)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (6.0.2)\n",
            "Requirement already satisfied: asynctest==0.13.0 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (0.13.0)\n",
            "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (4.0.2)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (1.8.1)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (22.1.0)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (1.3.1)\n",
            "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (2.1.1)\n",
            "Requirement already satisfied: idna>=2.0 in /usr/local/lib/python3.7/dist-packages (from yarl<2.0,>=1.0->aiohttp->gradio) (2.10)\n",
            "Requirement already satisfied: python-dateutil>2.1 in /usr/local/lib/python3.7/dist-packages (from analytics-python->gradio) (2.8.2)\n",
            "Requirement already satisfied: backoff==1.10.0 in /usr/local/lib/python3.7/dist-packages (from analytics-python->gradio) (1.10.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from analytics-python->gradio) (1.15.0)\n",
            "Requirement already satisfied: monotonic>=1.5 in /usr/local/lib/python3.7/dist-packages (from analytics-python->gradio) (1.6)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->gradio) (1.26.12)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->gradio) (2022.6.15)\n",
            "Requirement already satisfied: starlette==0.20.4 in /usr/local/lib/python3.7/dist-packages (from fastapi->gradio) (0.20.4)\n",
            "Requirement already satisfied: anyio<5,>=3.4.0 in /usr/local/lib/python3.7/dist-packages (from starlette==0.20.4->fastapi->gradio) (3.6.1)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.7/dist-packages (from anyio<5,>=3.4.0->starlette==0.20.4->fastapi->gradio) (1.3.0)\n",
            "Requirement already satisfied: rfc3986[idna2008]<2,>=1.3 in /usr/local/lib/python3.7/dist-packages (from httpx->gradio) (1.5.0)\n",
            "Requirement already satisfied: httpcore<0.16.0,>=0.15.0 in /usr/local/lib/python3.7/dist-packages (from httpx->gradio) (0.15.0)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.7/dist-packages (from jinja2->gradio) (2.0.1)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.7/dist-packages (from markdown-it-py[linkify,plugins]->gradio) (0.1.2)\n",
            "Requirement already satisfied: linkify-it-py~=1.0 in /usr/local/lib/python3.7/dist-packages (from markdown-it-py[linkify,plugins]->gradio) (1.0.3)\n",
            "Requirement already satisfied: mdit-py-plugins in /usr/local/lib/python3.7/dist-packages (from markdown-it-py[linkify,plugins]->gradio) (0.3.0)\n",
            "Requirement already satisfied: uc-micro-py in /usr/local/lib/python3.7/dist-packages (from linkify-it-py~=1.0->markdown-it-py[linkify,plugins]->gradio) (1.0.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib->gradio) (0.11.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->gradio) (1.4.4)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->gradio) (3.0.9)\n",
            "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.7/dist-packages (from pandas->gradio) (2022.2.1)\n",
            "Requirement already satisfied: cryptography>=2.5 in /usr/local/lib/python3.7/dist-packages (from paramiko->gradio) (38.0.1)\n",
            "Requirement already satisfied: pynacl>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from paramiko->gradio) (1.5.0)\n",
            "Requirement already satisfied: bcrypt>=3.1.3 in /usr/local/lib/python3.7/dist-packages (from paramiko->gradio) (4.0.0)\n",
            "Requirement already satisfied: cffi>=1.12 in /usr/local/lib/python3.7/dist-packages (from cryptography>=2.5->paramiko->gradio) (1.15.1)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.7/dist-packages (from cffi>=1.12->cryptography>=2.5->paramiko->gradio) (2.21)\n",
            "Requirement already satisfied: click>=7.0 in /usr/local/lib/python3.7/dist-packages (from uvicorn->gradio) (7.1.2)\n",
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: Bio in /usr/local/lib/python3.7/dist-packages (1.4.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.7/dist-packages (from Bio) (4.64.1)\n",
            "Requirement already satisfied: biopython>=1.79 in /usr/local/lib/python3.7/dist-packages (from Bio) (1.79)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from Bio) (2.28.1)\n",
            "Requirement already satisfied: mygene in /usr/local/lib/python3.7/dist-packages (from Bio) (3.2.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython>=1.79->Bio) (1.21.6)\n",
            "Requirement already satisfied: biothings-client>=0.2.6 in /usr/local/lib/python3.7/dist-packages (from mygene->Bio) (0.2.6)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (1.26.12)\n",
            "Requirement already satisfied: charset-normalizer<3,>=2 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.1.1)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2022.6.15)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.10)\n",
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: rmsd in /usr/local/lib/python3.7/dist-packages (1.4)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from rmsd) (1.21.6)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.7/dist-packages (from rmsd) (1.7.3)\n"
          ]
        }
      ],
      "source": [
        "#Loading all necessary modules/packages\n",
        "#Input protein sequence(s)\n",
        "from google.colab import files\n",
        "import os.path\n",
        "from os import path\n",
        "import re\n",
        "import hashlib\n",
        "import random\n",
        "\n",
        "#Run Prediction\n",
        "import sys\n",
        "import warnings\n",
        "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
        "\n",
        "#User Iterface\n",
        "import sys\n",
        "!{sys.executable} -m pip install gradio\n",
        "import gradio  as gr\n",
        "\n",
        "#RMSD Calculations\n",
        "!{sys.executable} -m pip install Bio\n",
        "!{sys.executable} -m pip install rmsd\n",
        "import Bio\n",
        "from Bio import PDB\n",
        "from pathlib import Path\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import shutil\n",
        "\n",
        "from colabfold.download import download_alphafold_params, default_data_dir\n",
        "from colabfold.utils import setup_logging\n",
        "from colabfold.batch import get_queries, run, set_model_type\n",
        "K80_chk = !nvidia-smi | grep \"Tesla K80\" | wc -l\n",
        "if \"1\" in K80_chk:\n",
        "  print(\"WARNING: found GPU Tesla K80: limited to total length < 1000\")\n",
        "  if \"TF_FORCE_UNIFIED_MEMORY\" in os.environ:\n",
        "    del os.environ[\"TF_FORCE_UNIFIED_MEMORY\"]\n",
        "  if \"XLA_PYTHON_CLIENT_MEM_FRACTION\" in os.environ:\n",
        "    del os.environ[\"XLA_PYTHON_CLIENT_MEM_FRACTION\"]\n",
        "\n",
        "from colabfold.colabfold import plot_protein\n",
        "from pathlib import Path\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "#Creates the directory path where the created PDB files are stored (if it doesnt exist already)\n",
        "if(not (os.path.isdir(\"PDBs\"))):\n",
        "  os.mkdir(\"PDBs\")\n",
        "\n",
        "if(not (os.path.isdir(\"aligned\"))):\n",
        "  os.mkdir(\"aligned\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "wAgO2BZE0Vml"
      },
      "outputs": [],
      "source": [
        "#Functional approach to iterating over given dictionary of values\n",
        "\n",
        "#Dictionary for TAG sequences\n",
        "tag_sequences = {\"6X His\":\"HHHHHH\", \"CBP\":\"KRRWKKNFIAVSAANRFKKISSSGAL\",\n",
        "                     \"CBD + 3G\": \"GGGTNPGVSAWQVNTAYTAGQLVTYNGKTYKCLQPHTSLAGWEPSNVPALWQLQ\",\n",
        "                     \"mNEONGreen + 5G\": \"GGGGGMVSKGEEDNMASLPATHELHIFGSINGVDFDMVGQGTGNPNDGYEELNLKSTKGDLQFSPWILVPHIGYGFHQYLPYPDGMSPFQAAMVDGSGYQVHRTMQFEDGASLTVNYRYTYEGSHIKGEAQVKGTGFPADGPVMTNSLTAADWCRSKKTYPNDKTIISTFKWSYTTGNGKRYRSTARTTYTFAKPMAANYLKNQPMYVFRKTELKHSKTELNFKEWQKAFTDVMGMDELYK\"\n",
        "                     }\n",
        "\n",
        "#Hash function for producing keys for appending to jobname\n",
        "def add_hash(x,y):\n",
        "  return x+\"_\"+hashlib.sha1(y.encode()).hexdigest()[:5]\n",
        "\n",
        "#Takes a dictionary of values and builds an array of all the filepaths\n",
        "def input_dict(variant_dict,isAdd):\n",
        "  #Clears PDBs files for analysis, so that new inputs can added\n",
        "  PDB_dir = \"/content/PDBs/\"\n",
        "  if (isAdd):\n",
        "    for PDB in os.listdir(PDB_dir):\n",
        "      os.remove(os.path.join(PDB_dir,PDB))\n",
        "\n",
        "  paths_dict = {}\n",
        "  for name in variant_dict:\n",
        "    # Initialization of the query_sequence\n",
        "    # Removes whitespaces\n",
        "    query_sequence = variant_dict[name]\n",
        "    query_sequence = \"\".join(query_sequence.split())\n",
        "    # Initialization of the jobname\n",
        "    # Removes whitespaces and append hash key\n",
        "    jobname = name\n",
        "    basejobname = \"\".join(jobname.split())\n",
        "    basejobname = re.sub(r'\\W+', '', basejobname)\n",
        "    jobname = add_hash(basejobname, query_sequence)\n",
        "    while os.path.isfile(f\"{jobname}.csv\"):\n",
        "      jobname = add_hash(basejobname, ''.join(random.sample(query_sequence,len(query_sequence))))\n",
        "\n",
        "    with open(f\"{jobname}.csv\", \"w\") as text_file:\n",
        "      text_file.write(f\"id,sequence\\n{jobname},{query_sequence}\")\n",
        "\n",
        "    #local variables needed to run prediction (most of these are adjustable, but for our purposes we are running them as default)\n",
        "    #MS\n",
        "    queries_path=f\"{jobname}.csv\"\n",
        "    use_amber = False\n",
        "    template_mode = \"none\"\n",
        "    custom_template_path = None\n",
        "    use_templates = False\n",
        "    msa_mode = \"MMseqs2 (UniRef+Environmental)\"\n",
        "    pair_mode = \"unpaired+paired\"\n",
        "    a3m_file = f\"{jobname}.a3m\"\n",
        "    model_type = \"auto\"\n",
        "    num_recycles = 3\n",
        "    save_to_google_drive = False\n",
        "    dpi = 200\n",
        "\n",
        "    input_array = [queries_path,use_amber,template_mode,custom_template_path,use_templates,msa_mode,pair_mode,a3m_file,model_type,num_recycles,save_to_google_drive,dpi]\n",
        "\n",
        "    if use_amber and '/usr/local/lib/python3.7/site-packages/' not in sys.path:\n",
        "      sys.path.insert(0, '/usr/local/lib/python3.7/site-packages/')\n",
        "\n",
        "    jobname_path = run_prediction(jobname,input_array,\"PDBs\")\n",
        "    paths_dict[name] = jobname_path\n",
        "\n",
        "  #Deletes all files stored in the sample data (to clear room for subsequent runs)\n",
        "  delete_dir = \"/content/\"\n",
        "  for file in os.listdir(delete_dir):\n",
        "    if(path.isfile(file)):\n",
        "      os.remove(os.path.join(delete_dir,file))\n",
        "\n",
        "  return paths_dict\n",
        "\n",
        "def prediction_callback(unrelaxed_protein, length, prediction_result, input_features, type):\n",
        "  fig = plot_protein(unrelaxed_protein, Ls=length, dpi=150)\n",
        "  plt.show()\n",
        "  plt.close()\n",
        "\n",
        "#Runs a prediction and moves the resulting rank 1 pdb structure (model does not matter) into a desired directory\n",
        "#Must provide the data array (contains all the parameters to run prediction) and directory path. \n",
        "#Returns the filepath of the PDB structure ofr the variant\n",
        "def run_prediction(jobname,data_array,directory_name):\n",
        "  #\n",
        "  result_dir=\".\"\n",
        "  if 'logging_setup' not in globals():\n",
        "      setup_logging(Path(\".\").joinpath(\"log.txt\"))\n",
        "      logging_setup = True\n",
        "  \n",
        "  queries_path=data_array[0]\n",
        "  use_amber = data_array[1]\n",
        "  template_mode = data_array[2]\n",
        "  custom_template_path = data_array[3]\n",
        "  use_templates = data_array[4]\n",
        "  msa_mode = data_array[5]\n",
        "  pair_mode = data_array[6]\n",
        "  a3m_file = data_array[7]\n",
        "  model_type = data_array[8]\n",
        "  num_recycles = data_array[9]\n",
        "  save_to_google_drive = data_array[10]\n",
        "  dpi = data_array[11]\n",
        "  \n",
        "  queries, is_complex = get_queries(queries_path)\n",
        "  model_type = set_model_type(is_complex, model_type)\n",
        "  download_alphafold_params(model_type, Path(\".\"))\n",
        "  run(\n",
        "      queries=queries,\n",
        "      result_dir=result_dir,\n",
        "      use_templates=use_templates,\n",
        "      custom_template_path=custom_template_path,\n",
        "      use_amber=use_amber,\n",
        "      msa_mode=msa_mode,    \n",
        "      model_type=model_type,\n",
        "      num_models=5,\n",
        "      num_recycles=num_recycles,\n",
        "      model_order=[1, 2, 3, 4, 5],\n",
        "      is_complex=is_complex,\n",
        "      data_dir=Path(\".\"),\n",
        "      keep_existing_results=False,\n",
        "      recompile_padding=1.0,\n",
        "      rank_by=\"auto\",\n",
        "      pair_mode=pair_mode,\n",
        "      stop_at_score=float(100),\n",
        "      prediction_callback=prediction_callback,\n",
        "      dpi=dpi\n",
        "  )\n",
        "\n",
        "  ex_end_path = \"\"\n",
        "\n",
        "  for index in range(1,6):\n",
        "    end_file = \"_unrelaxed_rank_1_model_\"+str(index)+\".pdb\"\n",
        "    test_path = \"/content/\" + jobname + end_file\n",
        "    if (os.path.exists(test_path)):\n",
        "      ex_end_path = end_file\n",
        "\n",
        "  old_path = \"/content/\" + jobname + ex_end_path\n",
        "  new_path = \"/content/\"+ directory_name +\"/\" + jobname + ex_end_path\n",
        "  os.replace(old_path,new_path)\n",
        "\n",
        "  return new_path"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "rCk6EPmgFPoW"
      },
      "outputs": [],
      "source": [
        "#Takes the given dictionaries for both natives and variant sequences, and performs RMSD calculations\n",
        "\n",
        "def process_tags(variant_dict):\n",
        "  keys = []\n",
        "  variant_dict.keys()\n",
        "  for key in variant_dict.keys():\n",
        "    if not key[9:] in keys:\n",
        "      keys.append(key[9:])\n",
        "  return keys\n",
        "\n",
        "def obtain_RMSD(native_dict,variant_dict):\n",
        "  data = process_tags(variant_dict)\n",
        "  for dir in os.listdir(\"/content/aligned/\"):\n",
        "    path = \"/content/aligned/\" + dir\n",
        "    shutil.rmtree(path)\n",
        "\n",
        "  native_names = native_dict.keys()\n",
        "  variant_names = variant_dict.keys()\n",
        "  for native_name in native_names:\n",
        "    parser = PDB.PDBParser()\n",
        "    ref_structure = parser.get_structure(native_name, native_dict[native_name])\n",
        "\n",
        "    model = ref_structure[0]\n",
        "    res_no= 0\n",
        "    non_resi = 0\n",
        "\n",
        "    for model in ref_structure:\n",
        "        for chain in model:\n",
        "            for r in chain.get_residues():\n",
        "                if r.id[0] == ' ':\n",
        "                    res_no +=1\n",
        "                else:\n",
        "                    non_resi +=1\n",
        "\n",
        "    res_no2 = 0\n",
        "    non_resi2 = 0\n",
        "    for model in ref_structure:\n",
        "      for residue in model.get_residues():\n",
        "          if PDB.is_aa(residue):\n",
        "              res_no2 += 1\n",
        "          else:\n",
        "              non_resi2 += 1\n",
        "\n",
        "    res_count = res_no\n",
        "\n",
        "    start_id = 1\n",
        "    end_id = res_no\n",
        "    atoms_to_be_aligned = range(start_id,end_id+1)\n",
        "    values=[]\n",
        "\n",
        "    align_dir = \"aligned/\" + native_name[:8]\n",
        "    os.mkdir(align_dir)\n",
        "\n",
        "    label_index = 0\n",
        "\n",
        "    for variant_name in variant_names:\n",
        "      if (variant_name[:8] == native_name[:8]):\n",
        "        ref_structure = parser.get_structure(variant_name, variant_dict[variant_name])\n",
        "        sample_structure = parser.get_structure(native_name, native_dict[native_name])\n",
        "      \n",
        "        ref_model    = ref_structure[0]\n",
        "        sample_model = sample_structure[0]\n",
        "\n",
        "        ref_atoms = []\n",
        "        sample_atoms = []\n",
        "\n",
        "         # Iterate of all chains in the model in order to find all residues\n",
        "        for ref_chain in ref_model:\n",
        "          # Iterate of all residues in each model in order to find proper atoms\n",
        "          for ref_res in ref_chain:\n",
        "            # Check if residue number ( .get_id() ) is in the list\n",
        "            if ref_res.get_id()[1] in atoms_to_be_aligned:\n",
        "              # Append CA atom to list\n",
        "              ref_atoms.append(ref_res['CA'])\n",
        "        \n",
        "        for sample_chain in sample_model:\n",
        "          for sample_res in sample_chain:\n",
        "            if sample_res.get_id()[1] in atoms_to_be_aligned:\n",
        "              sample_atoms.append(sample_res['CA'])\n",
        "\n",
        "        super_imposer = Bio.PDB.Superimposer()\n",
        "        super_imposer.set_atoms(ref_atoms, sample_atoms)\n",
        "        super_imposer.apply(sample_model.get_atoms())\n",
        "        values.append(super_imposer.rms)\n",
        "\n",
        "        io = Bio.PDB.PDBIO()\n",
        "        io.set_structure(sample_structure)\n",
        "        filename = native_name+\"_\"+data[label_index] + \"_aligned_.pdb\"\n",
        "        io.save(filename)\n",
        "        os.path.join(align_dir+\"/\",filename)\n",
        "        old_path = \"/content/\" + filename\n",
        "        new_path = \"/content/aligned/\" + native_name[:8] + \"/\" + filename\n",
        "        os.replace(old_path,new_path)\n",
        "        \n",
        "        label_index += 1\n",
        "      else:\n",
        "        continue\n",
        "\n",
        "  df = pd.DataFrame(data)\n",
        "  df.columns= ['tag']\n",
        "  #appending rmsd values to df\n",
        "  df['RMSD_values_proteinx']= values\n",
        "\n",
        "  x = df.loc[:, \"tag\"]\n",
        "  y = df.loc[:, \"RMSD_values_proteinx\"]\n",
        "\n",
        "  fig, ax = plt.subplots()\n",
        "  ax.stem(x,y)\n",
        "  plt.ylim([0,5])\n",
        "  # ax.set(xlim=(0,0),xticks= np.arrange(1,5),\n",
        "  #       ylim=(0,10), yticks= np.arrange(1,10))\n",
        "  return fig\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 646
        },
        "id": "mEC2_ISr41Ze",
        "outputId": "a2619600-a67c-44b7-e2d7-294d244fd964"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Colab notebook detected. To show errors in colab notebook, set `debug=True` in `launch()`\n",
            "Running on public URL: https://21725.gradio.app\n",
            "\n",
            "This share link expires in 72 hours. For free permanent hosting, check out Spaces: https://huggingface.co/spaces\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<div><iframe src=\"https://21725.gradio.app\" width=\"900\" height=\"500\" allow=\"autoplay; camera; microphone;\" frameborder=\"0\" allowfullscreen></iframe></div>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(<gradio.routes.App at 0x7f7b205ccd10>,\n",
              " 'http://127.0.0.1:7860/',\n",
              " 'https://21725.gradio.app')"
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ],
      "source": [
        "# Interface to determine number of proteins to compare where n is the number of proteins\n",
        "# 1 <= n <= 10\n",
        "\n",
        "files = {}\n",
        "\n",
        "def count_protein(quantity):\n",
        "    if (quantity == 1):\n",
        "        return gr.Textbox.update(lines=1, visible=True,placeholder=\"Enter protein sequence.\")\n",
        "    else:\n",
        "        return gr.Textbox.update(lines=quantity, visible=True,placeholder=\"Enter protein sequences separated by a semicolon\\nEnter a new sequence per line.\")\n",
        "\n",
        "def parse_choices(protein_num,seq_string,tag_choices):\n",
        "    filepaths_list = process_variants(process_str(seq_string),tag_choices)\n",
        "    native_list = input_dict(filepaths_list[0],True)\n",
        "    variant_list = input_dict(filepaths_list[1],False)\n",
        "    files = [native_list,variant_list]\n",
        "    plot = obtain_RMSD(native_list,variant_list)\n",
        "    return plot\n",
        "\n",
        "def process_str(seq_string):\n",
        "    seq_list = seq_string.split(\";\\n\")\n",
        "    return seq_list\n",
        "\n",
        "def process_variants(seq_list,tag_choices):\n",
        "    variant_list = {}\n",
        "    native_list = {}\n",
        "    for seq_index in range (0,len(seq_list)):\n",
        "        ori_seq = \"Protein\" + str(seq_index)\n",
        "        native_list[ori_seq] = seq_list[seq_index]\n",
        "        for tag in tag_choices:\n",
        "            variant_name = ori_seq + \"_\" + tag\n",
        "            variant_seq = seq_list[seq_index] + tag_sequences[tag]\n",
        "            variant_list[variant_name] = variant_seq\n",
        "    filepaths = [native_list,variant_list]\n",
        "    return filepaths\n",
        "        \n",
        "\n",
        "with gr.Blocks() as demo:\n",
        "    gr.Markdown(\n",
        "    \"\"\"\n",
        "    # Protein TAG Analyzer\n",
        "    Configure below:\n",
        "    \"\"\")\n",
        "    slider = gr.Slider(1,10,1,step=1,label=\"Number of Protein(s)\")\n",
        "    text = gr.Textbox(lines=1, interactive=True,placeholder=\"Enter protein sequence.\",label=\"Protein Sequence(s)\")\n",
        "    checkgroup = gr.CheckboxGroup([\"6X His\",\"CBP\",\"CBD + 3G\",\"mNEONGreen + 5G\"],label=\"Select TAGs\")\n",
        "    generate = gr.Button()\n",
        "    with gr.Row():\n",
        "            with gr.Column(scale=1):\n",
        "                rmsd_plot = gr.Plot()\n",
        "    \n",
        "    slider.change(fn=count_protein,inputs=slider,outputs=text)\n",
        "    generate.click(fn=parse_choices,inputs=[slider,text,checkgroup],outputs=rmsd_plot)\n",
        "\n",
        "demo.launch()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "files = [{'Protein0': '/content/PDBs/Protein0_9d99d_unrelaxed_rank_1_model_3.pdb'}, {'Protein0_CBP': '/content/PDBs/Protein0_CBP_f392e_unrelaxed_rank_1_model_5.pdb', 'Protein0_6X His': '/content/PDBs/Protein0_6XHis_89f0d_unrelaxed_rank_1_model_1.pdb'}]\n",
        "\n",
        "native_dict = files[0]\n",
        "variant_dict = files[1]\n",
        "\n",
        "obtain_RMSD(native_dict,variant_dict)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 559
        },
        "id": "81Yzx26QmVHb",
        "outputId": "a0dea014-6623-44a0-f4dd-f1159ca1192b"
      },
      "execution_count": 148,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:108: UserWarning: In Matplotlib 3.3 individual lines on a stem plot will be added as a LineCollection instead of individual lines. This significantly improves the performance of a stem plot. To remove this warning and switch to the new behaviour, set the \"use_line_collection\" keyword argument to True.\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWsAAAD8CAYAAACxUoU3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAALp0lEQVR4nO3df6zdd13H8dfbtmqd4BK5gOsIQ8QuSGB1N+CCGpnRTSWICTEjARU1nSbgiKbMERNjTIxaY8CoaBWERMYwc8xJkLqELWMRN9t1v2cDm5OtY9klUmHk6rry9o97ytrS3nva3nPXj308kqb3fM/nnPP+69nv/ZzvOa3uDgCnt296tgcAYGViDTAAsQYYgFgDDECsAQYg1gADWD/Noqp6OMlXkhxM8nR3z89yKACONFWsJ17X3V+c2SQAHJdtEIAB1DSfYKyq/0jypSSd5C+7e8cx1mxNsjVJzjrrrAvPP//8VR4V4P+v3bt3f7G75453/7Sx3tTd+6rq+UluTPKO7r7leOvn5+d7165dJzUwwJmoqnYv937gVNsg3b1v8vcTST6W5NWrMx4A01gx1lV1VlU959DPSX48yb2zHgyAZ0xzNcgLknysqg6tv7q7PznTqQA4woqx7u6HkrxqDWYB4DhcugcwALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMYOpYV9W6qtpTVR+f5UAAfKMTObO+IskDsxoEgOObKtZVdW6Sn0ry17MdB4BjmfbM+j1J3pXka8dbUFVbq2pXVe1aWFhYleEAWLJirKvq9Ume6O7dy63r7h3dPd/d83Nzc6s2IADTnVm/NskbqurhJNckubiq/namUwFwhBVj3d1Xdfe53X1eksuSfKq73zLzyQD4OtdZAwxg/Yks7u6bk9w8k0kAOC5n1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGsGOuq+taqur2q7qqq+6rqd9ZiMACesX6KNf+b5OLufrKqNiS5tar+qbv/dcazATCxYqy7u5M8Obm5YfKnZzkUAEeaas+6qtZV1Z1JnkhyY3ffdow1W6tqV1XtWlhYWO05Ac5oU8W6uw929wVJzk3y6qp6xTHW7Oju+e6en5ubW+05Ac5oJ3Q1SHfvT3JTkktnMw4Ax7LinnVVzSU50N37q2pjkh9L8gerPcj1e/Zl+869eWz/Ys45e2O2XbI5b9yyabVfBmBI01wN8l1JPlRV67J0Jv533f3x1Rzi+j37ctV192TxwMEkyb79i7nqunuSRLABMt3VIHcn2TLLIbbv3Pv1UB+yeOBgtu/cK9YAOU0+wfjY/sUTOg5wpjktYn3O2RtP6DjAmea0iPW2SzZn44Z1RxzbuGFdtl2y+VmaCOD0Ms0bjDN3aF/6XdfenacOfi2bXA0CcITTItbJUrA/cvvnkyQfvfyiZ3kagNPLabENAsDyxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBrBirKvqRVV1U1XdX1X3VdUVazEYAM9YP8Wap5P8RnffUVXPSbK7qm7s7vtnPBsAEyueWXf3F7r7jsnPX0nyQJJNsx4MgGec0J51VZ2XZEuS245x39aq2lVVuxYWFlZnOgCSnECsq+rbk/x9knd295ePvr+7d3T3fHfPz83NreaMAGe8qWJdVRuyFOoPd/d1sx0JgKNNczVIJXl/kge6+49nPxIAR5vmzPq1Sd6a5OKqunPy5ydnPBcAh1nx0r3uvjVJrcEsAByHTzACDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxgmv98AIBlXL9nX7bv3JvH9i/mnLM3Ztslm/PGLav7tf9iDXAKrt+zL1ddd08WDxxMkuzbv5irrrsnSVY12LZBAE7B9p17vx7qQxYPHMz2nXtX9XXEGuAUPLZ/8YSOnyyxBjgF55y98YSOnyyxBjgF2y7ZnI0b1h1xbOOGddl2yeZVfR1vMAKcgkNvIr7r2rvz1MGvZZOrQQBOT2/csikfuf3zSZKPXn7RTF7DNgjAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAVox1VX2gqp6oqnvXYiAAvtE0Z9YfTHLpjOcAYBkrxrq7b0nyX2swCwDHsWp71lW1tap2VdWuhYWF1XpaALKKse7uHd09393zc3Nzq/W0AMTVIABDEGuAAUxz6d5HknwmyeaqerSqfmn2YwFwuPUrLejuN6/FIAAcn20QgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABTxbqqLq2qvVX1uar6zVkPBcCRVox1Va1L8mdJfiLJy5O8uapePuvBAHjG+inWvDrJ57r7oSSpqmuS/HSS+1d7mEtvvjovXHgk/3nrc1f7qQFm6he+8OU8Pvei5PKLZvL808R6U5JHDrv9aJLXHL2oqrYm2Tq5+WRV7T3JmZ6X5Isn+ViAZ9PzfvGaPz3Zfr14uTunifVUuntHkh2n+jxVtau751dhJIA1Nct+TfMG474kLzrs9rmTYwCskWli/W9JXlZVL6mqb05yWZIbZjsWAIdbcRuku5+uqrcn2ZlkXZIPdPd9M5zplLdSAJ4lM+tXdfesnhuAVeITjAADEGuAAaxprKvqhVV1TVU9WFW7q+oTVfW9VbVYVXdW1V1V9S9VtXmy/keq6r8n9z1QVb+9lvMCZ56qOruqrq2qf59056LJ8Ruq6ucOW/dXVbXtGI//YFW96ahjT07+Pqeqrj2ZuVbtOuuVVFUl+ViSD3X3ZZNjr0rygiQPdvcFk2OXJ3l3kp+fPPTT3f36qjoryZ1V9Y/dfcdazQ2ccd6b5JPd/abJFXDfNjn+a0luqqobsvTVG69J8qsn8sTd/ViSN6248BjW8sz6dUkOdPdfHDrQ3XflyE9HJslzk3zp6Ad391eT7E7yPbMcEjhzVdV3JPnhJO9Pku5+qrv3T35+OEtXe/xhkvcleXt3P32Cz39eVd07+fn7qur2yc7B3VX1suUeu2Zn1klekaXYHstLq+rOJM/J0r9ix/o4+3cm+YEkvzuzCYEz3UuSLCT5m8lv/ruTXDE5WUySP0ryYJZ+479lmefZXlW/tcJr/UqS93b3hydn8OuWW3y6vMH4YHdf0N0vTfLOHHmt4g9V1Z4k/5zk92d8jTdwZluf5PuTvK+7tyT5apLDvxb6lVnq5vlVtVw/t02adsGhLd5j+EySd1fVlUle3N2Lyw22lrG+L8mFU6y7IUu/hhzy6e7e0t0XHr6FAjADjyZ5tLtvm9y+NkvxziTOf57kLUk+mxPcrz5ad1+d5A1JFpN8oqouXm79Wsb6U0m+ZfLtfEmSqnpljvzekST5wSz9mgGwprr78SSPHLoiLcmP5pmvg748yWe7++Ykv57kyqqaO9nXqqrvTvJQd/9Jkn/I0ln7ca3ZnnV3d1X9TJL3TE77/yfJw1na9ji0Z11Jnkryy2s1F8BR3pHk0D7yQ0neVlXPT3Jllt43S3c/VlXvydKbjW87ydf52SRvraoDSR5P8nvLLfZxc4ABnC5vMAKwDLEGGIBYAwxArAEGINYAAxBrgAGINcAA/g/Hn5Sn2B71DAAAAABJRU5ErkJggg==\n"
          },
          "metadata": {},
          "execution_count": 148
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWsAAAD8CAYAAACxUoU3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAALp0lEQVR4nO3df6zdd13H8dfbtmqd4BK5gOsIQ8QuSGB1N+CCGpnRTSWICTEjARU1nSbgiKbMERNjTIxaY8CoaBWERMYwc8xJkLqELWMRN9t1v2cDm5OtY9klUmHk6rry9o97ytrS3nva3nPXj308kqb3fM/nnPP+69nv/ZzvOa3uDgCnt296tgcAYGViDTAAsQYYgFgDDECsAQYg1gADWD/Noqp6OMlXkhxM8nR3z89yKACONFWsJ17X3V+c2SQAHJdtEIAB1DSfYKyq/0jypSSd5C+7e8cx1mxNsjVJzjrrrAvPP//8VR4V4P+v3bt3f7G75453/7Sx3tTd+6rq+UluTPKO7r7leOvn5+d7165dJzUwwJmoqnYv937gVNsg3b1v8vcTST6W5NWrMx4A01gx1lV1VlU959DPSX48yb2zHgyAZ0xzNcgLknysqg6tv7q7PznTqQA4woqx7u6HkrxqDWYB4DhcugcwALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMYOpYV9W6qtpTVR+f5UAAfKMTObO+IskDsxoEgOObKtZVdW6Sn0ry17MdB4BjmfbM+j1J3pXka8dbUFVbq2pXVe1aWFhYleEAWLJirKvq9Ume6O7dy63r7h3dPd/d83Nzc6s2IADTnVm/NskbqurhJNckubiq/namUwFwhBVj3d1Xdfe53X1eksuSfKq73zLzyQD4OtdZAwxg/Yks7u6bk9w8k0kAOC5n1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGsGOuq+taqur2q7qqq+6rqd9ZiMACesX6KNf+b5OLufrKqNiS5tar+qbv/dcazATCxYqy7u5M8Obm5YfKnZzkUAEeaas+6qtZV1Z1JnkhyY3ffdow1W6tqV1XtWlhYWO05Ac5oU8W6uw929wVJzk3y6qp6xTHW7Oju+e6en5ubW+05Ac5oJ3Q1SHfvT3JTkktnMw4Ax7LinnVVzSU50N37q2pjkh9L8gerPcj1e/Zl+869eWz/Ys45e2O2XbI5b9yyabVfBmBI01wN8l1JPlRV67J0Jv533f3x1Rzi+j37ctV192TxwMEkyb79i7nqunuSRLABMt3VIHcn2TLLIbbv3Pv1UB+yeOBgtu/cK9YAOU0+wfjY/sUTOg5wpjktYn3O2RtP6DjAmea0iPW2SzZn44Z1RxzbuGFdtl2y+VmaCOD0Ms0bjDN3aF/6XdfenacOfi2bXA0CcITTItbJUrA/cvvnkyQfvfyiZ3kagNPLabENAsDyxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBrBirKvqRVV1U1XdX1X3VdUVazEYAM9YP8Wap5P8RnffUVXPSbK7qm7s7vtnPBsAEyueWXf3F7r7jsnPX0nyQJJNsx4MgGec0J51VZ2XZEuS245x39aq2lVVuxYWFlZnOgCSnECsq+rbk/x9knd295ePvr+7d3T3fHfPz83NreaMAGe8qWJdVRuyFOoPd/d1sx0JgKNNczVIJXl/kge6+49nPxIAR5vmzPq1Sd6a5OKqunPy5ydnPBcAh1nx0r3uvjVJrcEsAByHTzACDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxgmv98AIBlXL9nX7bv3JvH9i/mnLM3Ztslm/PGLav7tf9iDXAKrt+zL1ddd08WDxxMkuzbv5irrrsnSVY12LZBAE7B9p17vx7qQxYPHMz2nXtX9XXEGuAUPLZ/8YSOnyyxBjgF55y98YSOnyyxBjgF2y7ZnI0b1h1xbOOGddl2yeZVfR1vMAKcgkNvIr7r2rvz1MGvZZOrQQBOT2/csikfuf3zSZKPXn7RTF7DNgjAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAVox1VX2gqp6oqnvXYiAAvtE0Z9YfTHLpjOcAYBkrxrq7b0nyX2swCwDHsWp71lW1tap2VdWuhYWF1XpaALKKse7uHd09393zc3Nzq/W0AMTVIABDEGuAAUxz6d5HknwmyeaqerSqfmn2YwFwuPUrLejuN6/FIAAcn20QgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABiDTAAsQYYgFgDDECsAQYg1gADEGuAAYg1wADEGmAAYg0wALEGGIBYAwxArAEGINYAAxBrgAGINcAAxBpgAGINMACxBhiAWAMMQKwBBiDWAAMQa4ABiDXAAMQaYABTxbqqLq2qvVX1uar6zVkPBcCRVox1Va1L8mdJfiLJy5O8uapePuvBAHjG+inWvDrJ57r7oSSpqmuS/HSS+1d7mEtvvjovXHgk/3nrc1f7qQFm6he+8OU8Pvei5PKLZvL808R6U5JHDrv9aJLXHL2oqrYm2Tq5+WRV7T3JmZ6X5Isn+ViAZ9PzfvGaPz3Zfr14uTunifVUuntHkh2n+jxVtau751dhJIA1Nct+TfMG474kLzrs9rmTYwCskWli/W9JXlZVL6mqb05yWZIbZjsWAIdbcRuku5+uqrcn2ZlkXZIPdPd9M5zplLdSAJ4lM+tXdfesnhuAVeITjAADEGuAAaxprKvqhVV1TVU9WFW7q+oTVfW9VbVYVXdW1V1V9S9VtXmy/keq6r8n9z1QVb+9lvMCZ56qOruqrq2qf59056LJ8Ruq6ucOW/dXVbXtGI//YFW96ahjT07+Pqeqrj2ZuVbtOuuVVFUl+ViSD3X3ZZNjr0rygiQPdvcFk2OXJ3l3kp+fPPTT3f36qjoryZ1V9Y/dfcdazQ2ccd6b5JPd/abJFXDfNjn+a0luqqobsvTVG69J8qsn8sTd/ViSN6248BjW8sz6dUkOdPdfHDrQ3XflyE9HJslzk3zp6Ad391eT7E7yPbMcEjhzVdV3JPnhJO9Pku5+qrv3T35+OEtXe/xhkvcleXt3P32Cz39eVd07+fn7qur2yc7B3VX1suUeu2Zn1klekaXYHstLq+rOJM/J0r9ix/o4+3cm+YEkvzuzCYEz3UuSLCT5m8lv/ruTXDE5WUySP0ryYJZ+479lmefZXlW/tcJr/UqS93b3hydn8OuWW3y6vMH4YHdf0N0vTfLOHHmt4g9V1Z4k/5zk92d8jTdwZluf5PuTvK+7tyT5apLDvxb6lVnq5vlVtVw/t02adsGhLd5j+EySd1fVlUle3N2Lyw22lrG+L8mFU6y7IUu/hhzy6e7e0t0XHr6FAjADjyZ5tLtvm9y+NkvxziTOf57kLUk+mxPcrz5ad1+d5A1JFpN8oqouXm79Wsb6U0m+ZfLtfEmSqnpljvzekST5wSz9mgGwprr78SSPHLoiLcmP5pmvg748yWe7++Ykv57kyqqaO9nXqqrvTvJQd/9Jkn/I0ln7ca3ZnnV3d1X9TJL3TE77/yfJw1na9ji0Z11Jnkryy2s1F8BR3pHk0D7yQ0neVlXPT3Jllt43S3c/VlXvydKbjW87ydf52SRvraoDSR5P8nvLLfZxc4ABnC5vMAKwDLEGGIBYAwxArAEGINYAAxBrgAGINcAA/g/Hn5Sn2B71DAAAAABJRU5ErkJggg==\n"
          },
          "metadata": {
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "gR8CUxXo8RAW"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "collapsed_sections": [],
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.13"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}