{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyOkPQG9hRDx9N3rZXEn0T2P",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/c-kong-ucsf/TAGScreeningProject/blob/main/dev_consolidated.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "5Sf6iXnzMV4a"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 901
        },
        "id": "4v8mrwrAlWT-",
        "outputId": "bb5b4b7d-2d47-4a53-ca75-c694d117d6b3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n",
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: Bio in /usr/local/lib/python3.7/dist-packages (1.4.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from Bio) (2.23.0)\n",
            "Requirement already satisfied: biopython>=1.79 in /usr/local/lib/python3.7/dist-packages (from Bio) (1.79)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.7/dist-packages (from Bio) (4.64.1)\n",
            "Requirement already satisfied: mygene in /usr/local/lib/python3.7/dist-packages (from Bio) (3.2.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython>=1.79->Bio) (1.21.6)\n",
            "Requirement already satisfied: biothings-client>=0.2.6 in /usr/local/lib/python3.7/dist-packages (from mygene->Bio) (0.2.6)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2022.6.15)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (3.0.4)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (1.24.3)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.10)\n",
            "Residues:  749\n",
            "Other:     0\n",
            "Residues2: 749\n",
            "Other2:    0\n",
            "749 amino acids selected for alignment and RMSD calculation\n",
            "2.327395039696388\n",
            "5.561542807408106\n",
            "4.219305812849976\n",
            "3.878207090476452\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:129: UserWarning: In Matplotlib 3.3 individual lines on a stem plot will be added as a LineCollection instead of individual lines. This significantly improves the performance of a stem plot. To remove this warning and switch to the new behaviour, set the \"use_line_collection\" keyword argument to True.\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-8-bedf6c82c943>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    128\u001b[0m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    129\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 130\u001b[0;31m ax.set(xlim=(0,0),xticks= np.arrange(1,5),\n\u001b[0m\u001b[1;32m    131\u001b[0m        ylim=(0,10), yticks= np.arrange(1,10))\n",
            "\u001b[0;31mNameError\u001b[0m: name 'np' is not defined"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAD4CAYAAADFAawfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAOUElEQVR4nO3df2ychX3H8c9nTgIOhUUdFm0CqhntXLVVwWDRRmWIsm6hDLWZtGl0aze2dYmmsVJtcmjWTdN+SVvdTXRax5TSlkoDypSFiEWsLmtBlP0AnDgkJMHjx/gRhyrHOo+28prE/u6Pe5zY3gU/ju+5+yb3fkkR5+ceP/naj3jnueeeu3NECACQ1w+1ewAAwOsj1ACQHKEGgOQINQAkR6gBILllVWz0/PPPj97e3io2DQBnpJ07d74aET2N7qsk1L29vRoZGali0wBwRrL94snu49QHACRHqAEgOUINAMkRagBIjlADQHKVXPWBM8P20XENDY/p0MSkVq/q1uC6Pq3vX9PusYCOQ6jR0PbRcW3etleTR6ckSeMTk9q8ba8kEWugxTj1gYaGhseOR3rG5NEpDQ2PtWkioHMRajR0aGJyUcsBVIdQo6HVq7oXtRxAdQg1Ghpc16fu5V1zlnUv79Lgur42TQR0Lp5MREMzTxhu2rpHR6amtYarPoC2IdQ4qfX9a3TP4y9Jku7duLbN0wCdi1MfAJAcoQaA5Ag1ACRHqAEgOUINAMkRagBIjlADQHKEGgCSI9QAkByhBoDkCDUAJEeoASC5Um/KZPsFSd+VNCXpWEQMVDkUAOCExbx73vsj4tXKJgEANMSpDwBIrmyoQ9LXbe+0vaHRCrY32B6xPVKr1Zo3IQB0uLKhvioiLpf0QUm/afvq+StExJaIGIiIgZ6enqYOCQCdrFSoI2K8+O9hSfdJurLKoQAAJywYatvn2D535rakn5L0VNWDAQDqylz1cYGk+2zPrH93RHyt0qkAAMctGOqIeF7SpS2YBQDQAJfnAUByhBoAkiPUAJAcoQaA5Ag1ACRHqAEgOUINAMkRagBIjlADQHKEGgCSI9QAkByhBoDkCDUAJEeoASA5Qg0AyRFqAEiOUANAcoQaAJIj1ACQXJkPtwWQyPbRcQ0Nj+nQxKRWr+rW4Lo+re9f0+6xUCFCDZxGto+Oa/O2vZo8OiVJGp+Y1OZteyWJWJ/BOPUBnEaGhseOR3rG5NEpDQ2PtWkitAKhBk4jhyYmF7UcZwZCDZxGVq/qXtRynBkINXAaGVzXp+7lXXOWdS/v0uC6vjZNhFbgyUTgNDLzhOGmrXt0ZGpaa7jqoyMQauA0s75/je55/CVJ0r0b17Z5GrRC6VMftrtsj9reUeVAAIC5FnOO+hZJB6oaBADQWKlQ275Q0k9LuqPacQAA85U9or5N0iZJ0ydbwfYG2yO2R2q1WlOGAwCUeDLR9g2SDkfETtvXnGy9iNgiaYskDQwMRNMmBIDkqn7/lTJXfbxP0odsXy/pbEnn2f67iPho06YAgNNUK95/ZcFTHxGxOSIujIheSTdK+iaRBoC6Vrz/Cq9MBIAlaMX7rywq1BHxcETc0LS/HQBOc614/xWOqAFgCVrx/iu8hBwAlqAV779CqAFgiap+/xVOfQBAcoQaAJIj1ACQHKEGgOQINQAkR6gBIDlCDQDJEWoASI5QA0ByhBoAkiPUAJAcoQaA5Ag1ACRHqAEgOUINAMkRagBIjlADQHKEGgCSI9QAkByhBoDkCDUAJEeoASA5Qg0AyS0Yattn237c9pO299n+w1YMBgCoW1ZinR9IujYivmd7uaRHbf9TRPx7xbMBAFQi1BERkr5XfLm8+BNVDgUAOKHUOWrbXbZ3Szos6cGIeKzBOhtsj9geqdVqzZ4TADpWqVBHxFREXCbpQklX2n5Xg3W2RMRARAz09PQ0e04A6FiLuuojIiYkPSTpumrGAQDMV+aqjx7bq4rb3ZJ+UtLTVQ8GAKgrc9XHmyV9xXaX6mH/+4jYUe1YAIAZZa762COpvwWzAAAa4JWJAJAcoQaA5Ag1ACRHqAEgOUINAMkRagBIjlADQHKEGgCSI9QAkByhBoDkCDUAJEeoASA5Qg0AyRFqAEiOUANAcmU+OKAlto+Oa2h4TIcmJrV6VbcG1/Vpff+ado8FAG2XItTbR8e1edteTR6dkiSNT0xq87a9kkSsAXS8FKc+hobHjkd6xuTRKQ0Nj7VpIgDII0WoD01MLmo5AHSSFKFevap7UcsBoJOkCPXguj51L++as6x7eZcG1/W1aSIAyCPFk4kzTxhu2rpHR6amtYarPgDguBShluqxvufxlyRJ925c2+ZpACCPFKc+AAAnR6gBIDlCDQDJEWoASG7BUNu+yPZDtvfb3mf7llYMBgCoK3PVxzFJvxMRu2yfK2mn7QcjYn/FswEAVOKIOiJeiYhdxe3vSjogiQucAaBFFnWO2navpH5JjzW4b4PtEdsjtVqtOdMBAMqH2vYbJP2DpE9GxGvz74+ILRExEBEDPT09zZwRADpaqVDbXq56pO+KiG3VjgQAmK3MVR+W9EVJByLiL6sfCQAwW5kj6vdJ+pika23vLv5cX/FcAIDCgpfnRcSjktyCWQAADfDKRABIjlADQHKEGgCSI9QAkByhBoDkCDUAJEeoASA5Qg0AyRFqAEiOUANAcoQaAJIj1ACQHKEGgOQINQAkR6gBIDlCDQDJEWoASI5QA0ByhBoAkiPUAJAcoQaA5Ag1ACRHqAEgOUINAMkRagBIjlADQHILhtr2l2wftv1UKwYCAMxV5oj6TknXVTwHAOAkFgx1RDwi6TstmAUA0EDTzlHb3mB7xPZIrVZr1mYBoOM1LdQRsSUiBiJioKenp1mbBYCOx1UfAJAcoQaA5MpcnnePpH+T1Gf7oO1fq34sAMCMZQutEBEfacUgAIDGOPUBAMkRagBIjlADQHKEGgCSI9QAkByhBoDkCDUAJEeoASA5Qg0AyRFqAEiOUANAcoQaAJIj1ACQHKEGgOQINQAkR6gBIDlCDQDJEWoASI5QA0ByhBoAkiPUAJAcoQaA5Ag1ACRHqAEgOUINAMkRagBIjlADQHKlQm37Ottjtp+1/amqhwIAnLBgqG13Sfq8pA9Keoekj9h+R9WDAQDqlpVY50pJz0bE85Jk+6uSPixpf7OHue7hu/Wm2st68dHzmr1pnKKbXnlNktgnybBf8rnpldf07Z6LpI1rm77tMqc+1kh6edbXB4tlc9jeYHvE9kitVjulYd54zllauaLrlL4X1Vi5oot9khD7JZ+VK7r0xnPOqmTbZY6oS4mILZK2SNLAwECcyjY+fMdnmzUOmuQt7R4ADbFf8qlyn5Q5oh6XdNGsry8slgEAWqBMqJ+Q9DbbF9teIelGSfdXOxYAYMaCpz4i4pjtmyUNS+qS9KWI2Ff5ZAAASSXPUUfEA5IeqHgWAEADvDIRAJIj1ACQHKEGgOQINQAk54hTem3K62/Urkl68RS//XxJrzZxHCwd+yQn9ks+S9knb4mInkZ3VBLqpbA9EhED7Z4DJ7BPcmK/5FPVPuHUBwAkR6gBILmMod7S7gHw/7BPcmK/5FPJPkl3jhoAMFfGI2oAwCyEGgCSqzTUtlfZ3mr7adsHbK+1fb/tX5q1zhdsDxa3H7Y9MOu+XttPFbevsb2juH2B7R22n7S93zZvGLUEtt9k+6u2n7O90/YDtn/M9qTt3cXv+V9t9xXrX2P7f4r7Dtj+g3b/DGeSJeyP0eJDqB+xfUO7f45OYvsm29O23z1r2VO2e4vbb7B9e7FPdxX79dfLbr9pn/ByEp+T9LWI+NnivaxXSvqEpIds36/6h+W+R9JvLHK7fyTpwYj4nCTN/uVgcWxb0n2SvhIRNxbLLpV0gaTnIuKyYtlGSb8r6ZeLb/1WRNxg+xxJu23/Y0Tsav1PcGZZ6v4o7rtM0nbbkxHxjVb/DB3soKRPS/r5BvfdIel5SW+LiGnbPZJ+teyGKzuitv3Dkq6W9EVJiogjETERES+o/szoZyTdLunmiDi2yM2/WfVfiopt72nK0J3p/ZKORsTfziyIiCc193MyJek8Sf89/5sj4vuSdkp6a5VDdpAl7Y9i/d2qH8zcXNWQnaR4ZP+07Ttt/4ftu2x/wPa/2H7G9pXFqjskvXPmkc6s779E9Q8J/72ImJakiKhFxJ+XnaHKI+qLJdUkfbk4Itgp6Zbif+zPSnpO9aOAR+Z93122J4vbKyRNN9j25yXdW3ygwT9L+nJEHKrih+gA71J93zRyie3dks5V/dHQe+avYPtHJL1X0h9XNmFnWdL+mGWXpMEmz9bJ3irp51Q/Cn5C0i9IukrSh1R/ZLNd9VZ9RnMf6UjSOyU9ORPpU1HlOeplki6XdHtE9Ev6vqRPFfe9u/i73257/gy/GBGXFQ/xrm+04YgYlvSjkr4g6e2SRouHEmiu54p9cYmkT2ruNaI/bntU0tcl/Rmf+tMSr7c/5nOLZuoU/xkRe4vY7pP0jahf27xXUu+s9e6W9F7bF59sQ7Y/XTzXUPrgsspQH5R0MCIeK77eKunyIsx/I+mjkp7R4s9PS5Ii4jsRcXdEfEz1f+GubsLMnWifpCtKrHe/5v6OvxUR/RFxxeyH6ViyU90f8/VLOtCUiSBJP5h1e3rW19OadWaiOI37F5JunbX+fkmXzhyURsSfFgei55X9yysLdUR8W9LLs87X/ITqA2+U9ExEPCzptyXdutijYdvX2l5Z3D5X0iWSXmrW7B3mm5LOsr1hZkHx5OxF89a7SvXTVajWkvdHsf7vq36KEK13p6QPSOqRpIh4VtKIpD+x3SVJts/WIh71VH3Vx2+pfs55herPeG6U9Jjq5zQVEYds36b6eZ1fWcR2r5D017aPqf6PzR0R8URTJ+8QERG2f0bSbbZvlfS/kl5Q/aH1zDlRSzoi6eNtG7RDLGF/zJyKWinpsKRPcMVHe0TEEdt/pfpVbzM+LmlI0rO2/0vSpKRNZbfJS8gBIDlemQgAyRFqAEiOUANAcoQaAJIj1ACQHKEGgOQINQAk93+pVdMA/htqnAAAAABJRU5ErkJggg==\n"
          },
          "metadata": {
            "needs_background": "light"
          }
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "#master install block\n",
        "!pip install Bio\n",
        "\n",
        "#master import block\n",
        "import Bio\n",
        "from Bio import PDB\n",
        "import numpy\n",
        "import pandas as pd\n",
        "import re\n",
        "import glob \n",
        "import os\n",
        "from pathlib import Path\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "##counting residues in default pdb(w/o tags) to determine which residues to run superimposition and subsequent RMSD measurement on in biopython\n",
        "parser= PDB.PDBParser()\n",
        "\n",
        "ref_structure = parser.get_structure(\"reference\", \"rank_1/a_4f680_unrelaxed_rank_1_model_3.pdb\")\n",
        "\n",
        "model = ref_structure[0]\n",
        "res_no= 0\n",
        "non_resi = 0\n",
        "\n",
        "for model in ref_structure:\n",
        "    for chain in model:\n",
        "        for r in chain.get_residues():\n",
        "            if r.id[0] == ' ':\n",
        "                res_no +=1\n",
        "            else:\n",
        "                non_resi +=1\n",
        "\n",
        "print (\"Residues:  %i\" % (res_no))\n",
        "print (\"Other:     %i\" % (non_resi))\n",
        "res_no2 = 0\n",
        "non_resi2 = 0\n",
        "for model in ref_structure:\n",
        "    for residue in model.get_residues():\n",
        "        if PDB.is_aa(residue):\n",
        "            res_no2 += 1\n",
        "\n",
        "        else:\n",
        "            non_resi2 += 1\n",
        "\n",
        "print (\"Residues2: %i\" % (res_no2))\n",
        "print (\"Other2:    %i\" % (non_resi2))\n",
        "res_count= res_no\n",
        "print(res_count,\"amino acids selected for alignment and RMSD calculation\")\n",
        "\n",
        "##running alignment  and subsequent MRSD calculation with Biopython (note: this code block predominantly came from https://gist.github.com/andersx/6354971)\n",
        "\n",
        "# Select what residues numbers you wish to align\n",
        "# and put them in a list\n",
        "start_id = 1\n",
        "end_id   = res_no\n",
        "atoms_to_be_aligned = range(start_id, end_id + 1)\n",
        "values=[]\n",
        "\n",
        "# for loop to iterate through tag PDB structures\n",
        "for filename in os.listdir('tagged_structures'):\n",
        "  f = 'tagged_structures/' + filename\n",
        "\n",
        "  # Get the structures\n",
        "  ref_structure = parser.get_structure(\"reference\", \"rank_1/a_4f680_unrelaxed_rank_1_model_3.pdb\")\n",
        "  sample_structure = parser.get_structure(\"sample\",f)\n",
        "\n",
        "  # Use the first model in the pdb-files for alignment\n",
        "  # Change the number 0 if you want to align to another structure\n",
        "  ref_model    = ref_structure[0]\n",
        "  sample_model = sample_structure[0]\n",
        "\n",
        "  # Make a list of the atoms (in the structures) you wish to align.\n",
        "  # In this case we use CA atoms whose index is in the specified range\n",
        "  ref_atoms = []\n",
        "  sample_atoms = []\n",
        "\n",
        "  # Iterate of all chains in the model in order to find all residues\n",
        "  for ref_chain in ref_model:\n",
        "    # Iterate of all residues in each model in order to find proper atoms\n",
        "    for ref_res in ref_chain:\n",
        "      # Check if residue number ( .get_id() ) is in the list\n",
        "      if ref_res.get_id()[1] in atoms_to_be_aligned:\n",
        "        # Append CA atom to list\n",
        "        ref_atoms.append(ref_res['CA'])\n",
        "\n",
        "  # Do the same for the sample structure\n",
        "  for sample_chain in sample_model:\n",
        "    for sample_res in sample_chain:\n",
        "      if sample_res.get_id()[1] in atoms_to_be_aligned:\n",
        "        sample_atoms.append(sample_res['CA'])\n",
        "\n",
        "  # Now we initiate the superimposer:\n",
        "  super_imposer = Bio.PDB.Superimposer()\n",
        "  super_imposer.set_atoms(ref_atoms, sample_atoms)\n",
        "  super_imposer.apply(sample_model.get_atoms())\n",
        "\n",
        "  # Print RMSD:\n",
        "  print(super_imposer.rms)\n",
        "  values.append(super_imposer.rms)\n",
        "\n",
        "\n",
        "  #a1_RMSD= super_imposer.rms\n",
        "\n",
        "  # Save the aligned version of 1UBQ.pdb\n",
        "  for i in range(1,5):      \n",
        "    io = Bio.PDB.PDBIO()\n",
        "    io.set_structure(sample_structure) \n",
        "    filename = \"a\" + str(i) + \"_aligned_.pdb\"\n",
        "    io.save(filename)\n",
        "    os.path.join(\"aligned/\",filename)\n",
        "\n",
        "#assembling pandas dataframe \n",
        "import pandas as pd\n",
        "data=['6XHIS', 'CBP', 'CBD', 'mNG']\n",
        "df = pd.DataFrame(data)\n",
        "df.columns= ['tag']\n",
        "#appending rmsd values to df\n",
        "df['RMSD_values_proteinx']= values\n",
        "\n",
        "#matplotlib-based plotting for tag variants of 1 protein\n",
        "import matplotlib.pyplot as plt\n",
        "x = df.loc[:, \"tag\"]\n",
        "y = df.loc[:, \"RMSD_values_proteinx\"]\n",
        "\n",
        "#plot\n",
        "fig, ax = plt.subplots()\n",
        "ax.stem(x,y)\n",
        "ax.set(xlim=(0,0),xticks= np.arrange(1,5),\n",
        "       ylim=(0,10), yticks= np.arrange(1,10))"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "ZWd8A5RWKkSS"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}