{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Gj7xA44e9bvA"
      },
      "source": [
        "##Intro to Programming Project: Protein TAG Variant Screen"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "cellView": "form",
        "id": "iccGdbe_Pmt9",
        "pycharm": {
          "name": "#%%\n"
        }
      },
      "outputs": [],
      "source": [
        "#@title Install dependencies\n",
        "%%bash -s $use_amber $use_templates\n",
        "\n",
        "set -e\n",
        "\n",
        "USE_AMBER=$1\n",
        "USE_TEMPLATES=$2\n",
        "\n",
        "if [ ! -f COLABFOLD_READY ]; then\n",
        "  # install dependencies\n",
        "  # We have to use \"--no-warn-conflicts\" because colab already has a lot preinstalled with requirements different to ours\n",
        "  pip install -q --no-warn-conflicts \"colabfold[alphafold-minus-jax] @ git+https://github.com/sokrypton/ColabFold\"\n",
        "  # high risk high gain\n",
        "  pip install -q \"jax[cuda11_cudnn805]>=0.3.8,<0.4\" -f https://storage.googleapis.com/jax-releases/jax_releases.html\n",
        "  touch COLABFOLD_READY\n",
        "fi\n",
        "\n",
        "# setup conda\n",
        "if [ ${USE_AMBER} == \"True\" ] || [ ${USE_TEMPLATES} == \"True\" ]; then\n",
        "  if [ ! -f CONDA_READY ]; then\n",
        "    wget -qnc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
        "    bash Miniconda3-latest-Linux-x86_64.sh -bfp /usr/local 2>&1 1>/dev/null\n",
        "    rm Miniconda3-latest-Linux-x86_64.sh\n",
        "    touch CONDA_READY\n",
        "  fi\n",
        "fi\n",
        "# setup template search\n",
        "if [ ${USE_TEMPLATES} == \"True\" ] && [ ! -f HH_READY ]; then\n",
        "  conda install -y -q -c conda-forge -c bioconda kalign2=2.04 hhsuite=3.3.0 python=3.7 2>&1 1>/dev/null\n",
        "  touch HH_READY\n",
        "fi\n",
        "# setup openmm for amber refinement\n",
        "if [ ${USE_AMBER} == \"True\" ] && [ ! -f AMBER_READY ]; then\n",
        "  conda install -y -q -c conda-forge openmm=7.5.1 python=3.7 pdbfixer 2>&1 1>/dev/null\n",
        "  touch AMBER_READY\n",
        "fi"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yE0sYroghJRg",
        "outputId": "881985a4-6073-467e-d42e-7ab778ee3883"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: gradio in /usr/local/lib/python3.7/dist-packages (3.3.1)\n",
            "Requirement already satisfied: orjson in /usr/local/lib/python3.7/dist-packages (from gradio) (3.8.0)\n",
            "Requirement already satisfied: pillow in /usr/local/lib/python3.7/dist-packages (from gradio) (7.1.2)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.7/dist-packages (from gradio) (6.0)\n",
            "Requirement already satisfied: h11<0.13,>=0.11 in /usr/local/lib/python3.7/dist-packages (from gradio) (0.12.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from gradio) (1.21.6)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.7/dist-packages (from gradio) (3.1.3)\n",
            "Requirement already satisfied: pydantic in /usr/local/lib/python3.7/dist-packages (from gradio) (1.9.2)\n",
            "Requirement already satisfied: pydub in /usr/local/lib/python3.7/dist-packages (from gradio) (0.25.1)\n",
            "Requirement already satisfied: ffmpy in /usr/local/lib/python3.7/dist-packages (from gradio) (0.3.0)\n",
            "Requirement already satisfied: pycryptodome in /usr/local/lib/python3.7/dist-packages (from gradio) (3.15.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from gradio) (2.28.1)\n",
            "Requirement already satisfied: uvicorn in /usr/local/lib/python3.7/dist-packages (from gradio) (0.18.3)\n",
            "Requirement already satisfied: websockets in /usr/local/lib/python3.7/dist-packages (from gradio) (10.3)\n",
            "Requirement already satisfied: python-multipart in /usr/local/lib/python3.7/dist-packages (from gradio) (0.0.5)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.7/dist-packages (from gradio) (2022.8.2)\n",
            "Requirement already satisfied: analytics-python in /usr/local/lib/python3.7/dist-packages (from gradio) (1.4.0)\n",
            "Requirement already satisfied: httpx in /usr/local/lib/python3.7/dist-packages (from gradio) (0.23.0)\n",
            "Requirement already satisfied: aiohttp in /usr/local/lib/python3.7/dist-packages (from gradio) (3.8.1)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.7/dist-packages (from gradio) (2.11.3)\n",
            "Requirement already satisfied: fastapi in /usr/local/lib/python3.7/dist-packages (from gradio) (0.85.0)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from gradio) (1.3.5)\n",
            "Requirement already satisfied: paramiko in /usr/local/lib/python3.7/dist-packages (from gradio) (2.11.0)\n",
            "Requirement already satisfied: markdown-it-py[linkify,plugins] in /usr/local/lib/python3.7/dist-packages (from gradio) (2.1.0)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (1.3.1)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (6.0.2)\n",
            "Requirement already satisfied: asynctest==0.13.0 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (0.13.0)\n",
            "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (4.0.2)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (1.8.1)\n",
            "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (2.1.1)\n",
            "Requirement already satisfied: typing-extensions>=3.7.4 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (4.1.1)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (1.2.0)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.7/dist-packages (from aiohttp->gradio) (22.1.0)\n",
            "Requirement already satisfied: idna>=2.0 in /usr/local/lib/python3.7/dist-packages (from yarl<2.0,>=1.0->aiohttp->gradio) (2.10)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from analytics-python->gradio) (1.15.0)\n",
            "Requirement already satisfied: python-dateutil>2.1 in /usr/local/lib/python3.7/dist-packages (from analytics-python->gradio) (2.8.2)\n",
            "Requirement already satisfied: monotonic>=1.5 in /usr/local/lib/python3.7/dist-packages (from analytics-python->gradio) (1.6)\n",
            "Requirement already satisfied: backoff==1.10.0 in /usr/local/lib/python3.7/dist-packages (from analytics-python->gradio) (1.10.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->gradio) (2022.6.15)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->gradio) (1.26.12)\n",
            "Requirement already satisfied: starlette==0.20.4 in /usr/local/lib/python3.7/dist-packages (from fastapi->gradio) (0.20.4)\n",
            "Requirement already satisfied: anyio<5,>=3.4.0 in /usr/local/lib/python3.7/dist-packages (from starlette==0.20.4->fastapi->gradio) (3.6.1)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.7/dist-packages (from anyio<5,>=3.4.0->starlette==0.20.4->fastapi->gradio) (1.3.0)\n",
            "Requirement already satisfied: httpcore<0.16.0,>=0.15.0 in /usr/local/lib/python3.7/dist-packages (from httpx->gradio) (0.15.0)\n",
            "Requirement already satisfied: rfc3986[idna2008]<2,>=1.3 in /usr/local/lib/python3.7/dist-packages (from httpx->gradio) (1.5.0)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.7/dist-packages (from jinja2->gradio) (2.0.1)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.7/dist-packages (from markdown-it-py[linkify,plugins]->gradio) (0.1.2)\n",
            "Requirement already satisfied: linkify-it-py~=1.0 in /usr/local/lib/python3.7/dist-packages (from markdown-it-py[linkify,plugins]->gradio) (1.0.3)\n",
            "Requirement already satisfied: mdit-py-plugins in /usr/local/lib/python3.7/dist-packages (from markdown-it-py[linkify,plugins]->gradio) (0.3.0)\n",
            "Requirement already satisfied: uc-micro-py in /usr/local/lib/python3.7/dist-packages (from linkify-it-py~=1.0->markdown-it-py[linkify,plugins]->gradio) (1.0.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib->gradio) (0.11.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->gradio) (1.4.4)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->gradio) (3.0.9)\n",
            "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.7/dist-packages (from pandas->gradio) (2022.2.1)\n",
            "Requirement already satisfied: bcrypt>=3.1.3 in /usr/local/lib/python3.7/dist-packages (from paramiko->gradio) (4.0.0)\n",
            "Requirement already satisfied: pynacl>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from paramiko->gradio) (1.5.0)\n",
            "Requirement already satisfied: cryptography>=2.5 in /usr/local/lib/python3.7/dist-packages (from paramiko->gradio) (38.0.1)\n",
            "Requirement already satisfied: cffi>=1.12 in /usr/local/lib/python3.7/dist-packages (from cryptography>=2.5->paramiko->gradio) (1.15.1)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.7/dist-packages (from cffi>=1.12->cryptography>=2.5->paramiko->gradio) (2.21)\n",
            "Requirement already satisfied: click>=7.0 in /usr/local/lib/python3.7/dist-packages (from uvicorn->gradio) (7.1.2)\n",
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: Bio in /usr/local/lib/python3.7/dist-packages (1.4.0)\n",
            "Requirement already satisfied: biopython>=1.79 in /usr/local/lib/python3.7/dist-packages (from Bio) (1.79)\n",
            "Requirement already satisfied: mygene in /usr/local/lib/python3.7/dist-packages (from Bio) (3.2.2)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.7/dist-packages (from Bio) (4.64.1)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from Bio) (2.28.1)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython>=1.79->Bio) (1.21.6)\n",
            "Requirement already satisfied: biothings-client>=0.2.6 in /usr/local/lib/python3.7/dist-packages (from mygene->Bio) (0.2.6)\n",
            "Requirement already satisfied: charset-normalizer<3,>=2 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.1.1)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2022.6.15)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.10)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (1.26.12)\n",
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: rmsd in /usr/local/lib/python3.7/dist-packages (1.4)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from rmsd) (1.21.6)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.7/dist-packages (from rmsd) (1.7.3)\n"
          ]
        }
      ],
      "source": [
        "#Loading all necessary modules/packages\n",
        "#Input protein sequence(s)\n",
        "from google.colab import files\n",
        "import os.path\n",
        "from os import path\n",
        "import re\n",
        "import hashlib\n",
        "import random\n",
        "\n",
        "#Run Prediction\n",
        "import sys\n",
        "import warnings\n",
        "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
        "\n",
        "#User Iterface\n",
        "import sys\n",
        "!{sys.executable} -m pip install gradio\n",
        "import gradio  as gr\n",
        "\n",
        "#RMSD Calculations\n",
        "!{sys.executable} -m pip install Bio\n",
        "!{sys.executable} -m pip install rmsd\n",
        "import Bio\n",
        "from Bio import PDB\n",
        "from pathlib import Path\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import shutil\n",
        "\n",
        "from colabfold.download import download_alphafold_params, default_data_dir\n",
        "from colabfold.utils import setup_logging\n",
        "from colabfold.batch import get_queries, run, set_model_type\n",
        "K80_chk = !nvidia-smi | grep \"Tesla K80\" | wc -l\n",
        "if \"1\" in K80_chk:\n",
        "  print(\"WARNING: found GPU Tesla K80: limited to total length < 1000\")\n",
        "  if \"TF_FORCE_UNIFIED_MEMORY\" in os.environ:\n",
        "    del os.environ[\"TF_FORCE_UNIFIED_MEMORY\"]\n",
        "  if \"XLA_PYTHON_CLIENT_MEM_FRACTION\" in os.environ:\n",
        "    del os.environ[\"XLA_PYTHON_CLIENT_MEM_FRACTION\"]\n",
        "\n",
        "from colabfold.colabfold import plot_protein\n",
        "from pathlib import Path\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "#Creates the directory path where the created PDB files are stored (if it doesnt exist already)\n",
        "if(not (os.path.isdir(\"PDBs\"))):\n",
        "  os.mkdir(\"PDBs\")\n",
        "\n",
        "if(not (os.path.isdir(\"aligned\"))):\n",
        "  os.mkdir(\"aligned\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "wAgO2BZE0Vml"
      },
      "outputs": [],
      "source": [
        "#Functional approach to iterating over given dictionary of values\n",
        "\n",
        "#Dictionary for TAG sequences\n",
        "tag_sequences = {\"6X His\":\"HHHHHH\", \"CBP\":\"KRRWKKNFIAVSAANRFKKISSSGAL\",\n",
        "                     \"CBD + 3G\": \"GGGTNPGVSAWQVNTAYTAGQLVTYNGKTYKCLQPHTSLAGWEPSNVPALWQLQ\",\n",
        "                     \"mNEONGreen + 5G\": \"GGGGGMVSKGEEDNMASLPATHELHIFGSINGVDFDMVGQGTGNPNDGYEELNLKSTKGDLQFSPWILVPHIGYGFHQYLPYPDGMSPFQAAMVDGSGYQVHRTMQFEDGASLTVNYRYTYEGSHIKGEAQVKGTGFPADGPVMTNSLTAADWCRSKKTYPNDKTIISTFKWSYTTGNGKRYRSTARTTYTFAKPMAANYLKNQPMYVFRKTELKHSKTELNFKEWQKAFTDVMGMDELYK\"\n",
        "                     }\n",
        "\n",
        "#Hash function for producing keys for appending to jobname\n",
        "def add_hash(x,y):\n",
        "  return x+\"_\"+hashlib.sha1(y.encode()).hexdigest()[:5]\n",
        "\n",
        "#Takes a dictionary of values and builds an array of all the filepaths\n",
        "def input_dict(variant_dict,isAdd):\n",
        "  #Clears PDBs files for analysis, so that new inputs can added\n",
        "  PDB_dir = \"/content/PDBs/\"\n",
        "  if (isAdd):\n",
        "    for PDB in os.listdir(PDB_dir):\n",
        "      os.remove(os.path.join(PDB_dir,PDB))\n",
        "\n",
        "  paths_dict = {}\n",
        "  for name in variant_dict:\n",
        "    # Initialization of the query_sequence\n",
        "    # Removes whitespaces\n",
        "    query_sequence = variant_dict[name]\n",
        "    query_sequence = \"\".join(query_sequence.split())\n",
        "    # Initialization of the jobname\n",
        "    # Removes whitespaces and append hash key\n",
        "    jobname = name\n",
        "    basejobname = \"\".join(jobname.split())\n",
        "    basejobname = re.sub(r'\\W+', '', basejobname)\n",
        "    jobname = add_hash(basejobname, query_sequence)\n",
        "    while os.path.isfile(f\"{jobname}.csv\"):\n",
        "      jobname = add_hash(basejobname, ''.join(random.sample(query_sequence,len(query_sequence))))\n",
        "\n",
        "    with open(f\"{jobname}.csv\", \"w\") as text_file:\n",
        "      text_file.write(f\"id,sequence\\n{jobname},{query_sequence}\")\n",
        "\n",
        "    #local variables needed to run prediction (most of these are adjustable, but for our purposes we are running them as default)\n",
        "    #MS\n",
        "    queries_path=f\"{jobname}.csv\"\n",
        "    use_amber = False\n",
        "    template_mode = \"none\"\n",
        "    custom_template_path = None\n",
        "    use_templates = False\n",
        "    msa_mode = \"MMseqs2 (UniRef+Environmental)\"\n",
        "    pair_mode = \"unpaired+paired\"\n",
        "    a3m_file = f\"{jobname}.a3m\"\n",
        "    model_type = \"auto\"\n",
        "    num_recycles = 3\n",
        "    save_to_google_drive = False\n",
        "    dpi = 200\n",
        "\n",
        "    input_array = [queries_path,use_amber,template_mode,custom_template_path,use_templates,msa_mode,pair_mode,a3m_file,model_type,num_recycles,save_to_google_drive,dpi]\n",
        "\n",
        "    if use_amber and '/usr/local/lib/python3.7/site-packages/' not in sys.path:\n",
        "      sys.path.insert(0, '/usr/local/lib/python3.7/site-packages/')\n",
        "\n",
        "    jobname_path = run_prediction(jobname,input_array,\"PDBs\")\n",
        "    paths_dict[name] = jobname_path\n",
        "\n",
        "  #Deletes all files stored in the sample data (to clear room for subsequent runs)\n",
        "  delete_dir = \"/content/\"\n",
        "  for file in os.listdir(delete_dir):\n",
        "    if(path.isfile(file)):\n",
        "      os.remove(os.path.join(delete_dir,file))\n",
        "\n",
        "  return paths_dict\n",
        "\n",
        "def prediction_callback(unrelaxed_protein, length, prediction_result, input_features, type):\n",
        "  fig = plot_protein(unrelaxed_protein, Ls=length, dpi=150)\n",
        "  plt.show()\n",
        "  plt.close()\n",
        "\n",
        "#Runs a prediction and moves the resulting rank 1 pdb structure (model does not matter) into a desired directory\n",
        "#Must provide the data array (contains all the parameters to run prediction) and directory path. \n",
        "#Returns the filepath of the PDB structure ofr the variant\n",
        "def run_prediction(jobname,data_array,directory_name):\n",
        "  #\n",
        "  result_dir=\".\"\n",
        "  if 'logging_setup' not in globals():\n",
        "      setup_logging(Path(\".\").joinpath(\"log.txt\"))\n",
        "      logging_setup = True\n",
        "  \n",
        "  queries_path=data_array[0]\n",
        "  use_amber = data_array[1]\n",
        "  template_mode = data_array[2]\n",
        "  custom_template_path = data_array[3]\n",
        "  use_templates = data_array[4]\n",
        "  msa_mode = data_array[5]\n",
        "  pair_mode = data_array[6]\n",
        "  a3m_file = data_array[7]\n",
        "  model_type = data_array[8]\n",
        "  num_recycles = data_array[9]\n",
        "  save_to_google_drive = data_array[10]\n",
        "  dpi = data_array[11]\n",
        "  \n",
        "  queries, is_complex = get_queries(queries_path)\n",
        "  model_type = set_model_type(is_complex, model_type)\n",
        "  download_alphafold_params(model_type, Path(\".\"))\n",
        "  run(\n",
        "      queries=queries,\n",
        "      result_dir=result_dir,\n",
        "      use_templates=use_templates,\n",
        "      custom_template_path=custom_template_path,\n",
        "      use_amber=use_amber,\n",
        "      msa_mode=msa_mode,    \n",
        "      model_type=model_type,\n",
        "      num_models=5,\n",
        "      num_recycles=num_recycles,\n",
        "      model_order=[1, 2, 3, 4, 5],\n",
        "      is_complex=is_complex,\n",
        "      data_dir=Path(\".\"),\n",
        "      keep_existing_results=False,\n",
        "      recompile_padding=1.0,\n",
        "      rank_by=\"auto\",\n",
        "      pair_mode=pair_mode,\n",
        "      stop_at_score=float(100),\n",
        "      prediction_callback=prediction_callback,\n",
        "      dpi=dpi\n",
        "  )\n",
        "\n",
        "  ex_end_path = \"\"\n",
        "\n",
        "  for index in range(1,6):\n",
        "    end_file = \"_unrelaxed_rank_1_model_\"+str(index)+\".pdb\"\n",
        "    test_path = \"/content/\" + jobname + end_file\n",
        "    if (os.path.exists(test_path)):\n",
        "      ex_end_path = end_file\n",
        "\n",
        "  old_path = \"/content/\" + jobname + ex_end_path\n",
        "  new_path = \"/content/\"+ directory_name +\"/\" + jobname + ex_end_path\n",
        "  os.replace(old_path,new_path)\n",
        "\n",
        "  return new_path"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 69,
      "metadata": {
        "id": "rCk6EPmgFPoW"
      },
      "outputs": [],
      "source": [
        "from pandas.core import frame\n",
        "#Takes the given dictionaries for both natives and variant sequences, and performs RMSD calculations\n",
        "\n",
        "def process_tags(variant_dict):\n",
        "  keys = []\n",
        "  variant_dict.keys()\n",
        "  for key in variant_dict.keys():\n",
        "    if not key[9:] in keys:\n",
        "      keys.append(key[9:])\n",
        "  return keys\n",
        "\n",
        "def obtain_RMSD(native_dict,variant_dict):\n",
        "  data = process_tags(variant_dict)\n",
        "  for dir in os.listdir(\"/content/aligned/\"):\n",
        "    path = \"/content/aligned/\" + dir\n",
        "    shutil.rmtree(path)\n",
        "\n",
        "  native_names = native_dict.keys()\n",
        "  variant_names = variant_dict.keys()\n",
        "  rmsd_dfs=[]\n",
        "  for native_name in native_names:\n",
        "    parser = PDB.PDBParser()\n",
        "    ref_structure = parser.get_structure(native_name, native_dict[native_name])\n",
        "\n",
        "    model = ref_structure[0]\n",
        "    res_no= 0\n",
        "    non_resi = 0\n",
        "\n",
        "    for model in ref_structure:\n",
        "        for chain in model:\n",
        "            for r in chain.get_residues():\n",
        "                if r.id[0] == ' ':\n",
        "                    res_no +=1\n",
        "                else:\n",
        "                    non_resi +=1\n",
        "\n",
        "    res_no2 = 0\n",
        "    non_resi2 = 0\n",
        "    for model in ref_structure:\n",
        "      for residue in model.get_residues():\n",
        "          if PDB.is_aa(residue):\n",
        "              res_no2 += 1\n",
        "          else:\n",
        "              non_resi2 += 1\n",
        "\n",
        "    res_count = res_no\n",
        "\n",
        "    start_id = 1\n",
        "    end_id = res_no\n",
        "    atoms_to_be_aligned = range(start_id,end_id+1)\n",
        "\n",
        "    align_dir = \"aligned/\" + native_name[:8]\n",
        "    os.mkdir(align_dir)\n",
        "\n",
        "    label_index = 0\n",
        "    values = []\n",
        "    for variant_name in variant_names:\n",
        "      if (variant_name[:8] == native_name[:8]):\n",
        "        ref_structure = parser.get_structure(variant_name, variant_dict[variant_name])\n",
        "        sample_structure = parser.get_structure(native_name, native_dict[native_name])\n",
        "      \n",
        "        ref_model    = ref_structure[0]\n",
        "        sample_model = sample_structure[0]\n",
        "\n",
        "        ref_atoms = []\n",
        "        sample_atoms = []\n",
        "\n",
        "         # Iterate of all chains in the model in order to find all residues\n",
        "        for ref_chain in ref_model:\n",
        "          # Iterate of all residues in each model in order to find proper atoms\n",
        "          for ref_res in ref_chain:\n",
        "            # Check if residue number ( .get_id() ) is in the list\n",
        "            if ref_res.get_id()[1] in atoms_to_be_aligned:\n",
        "              # Append CA atom to list\n",
        "              ref_atoms.append(ref_res['CA'])\n",
        "        \n",
        "        for sample_chain in sample_model:\n",
        "          for sample_res in sample_chain:\n",
        "            if sample_res.get_id()[1] in atoms_to_be_aligned:\n",
        "              sample_atoms.append(sample_res['CA'])\n",
        "\n",
        "        super_imposer = Bio.PDB.Superimposer()\n",
        "        super_imposer.set_atoms(ref_atoms, sample_atoms)\n",
        "        super_imposer.apply(sample_model.get_atoms())\n",
        "        values.append(super_imposer.rms)\n",
        "\n",
        "        io = Bio.PDB.PDBIO()\n",
        "        io.set_structure(sample_structure)\n",
        "        filename = native_name+\"_\"+data[label_index] + \"_aligned_.pdb\"\n",
        "        io.save(filename)\n",
        "        os.path.join(align_dir+\"/\",filename)\n",
        "        old_path = \"/content/\" + filename\n",
        "        new_path = \"/content/aligned/\" + native_name[:8] + \"/\" + filename\n",
        "        os.replace(old_path,new_path)\n",
        "        \n",
        "        label_index += 1\n",
        "      else:\n",
        "        continue\n",
        "    rmsd_dfs.append(values)\n",
        "\n",
        "  # x = np.arange(len(data))\n",
        "  # protein_index = 0\n",
        "  \n",
        "  # fig, ax = plt.subplots(figsize=(10,6))\n",
        "  # for protein in rmsd_dfs:\n",
        "  #   ax.bar(x,protein,width=.35)\n",
        "  \n",
        "  # ax.set_ylabel('RMSD values')\n",
        "  # ax.set_title('RMSD calculations for tagged vs native sequence')\n",
        "  # ax.set_xticks(x)\n",
        "  # ax.set_xticklabels(data)\n",
        "\n",
        "  # # for c in ax.containers:\n",
        "  # #   ax.bar_label(c)\n",
        "\n",
        "  # return plt\n",
        "\n",
        "  N = len(rmsd_dfs)\n",
        "  ind = np.arange(N)\n",
        "  print(ind)\n",
        "  width = 0\n",
        "\n",
        "  protein_index = 0\n",
        "  for protein in rmsd_dfs:\n",
        "    plt.bar(ind+width, protein, 0.4, label = \"Protein\"+str(protein_index))\n",
        "    protein_index += 1\n",
        "    width +=0.4 \n",
        "\n",
        "  plt.xticks(ind+0.4,data)\n",
        "  plt.xlabel(\"Tags\")\n",
        "  plt.ylabel(\"RMSD values\")\n",
        "  plt.title(\"RMSD calculations for tagged vs native sequence\")\n",
        "  return plt\n",
        "\n",
        "  #   for protein_df in rmsd_dfs:\n",
        "  # return rmsd_dfs\n",
        "\n",
        "  # fig = plt.figure()\n",
        "\n",
        "  # for frame in rmsd_dfs:\n",
        "  #   plt.plot(frame[\"Test\"],frame[\"RMSD\"])\n",
        "  \n",
        "  # return fig\n",
        "  # # df = pd.DataFrame(data)\n",
        "  # # df.columns= ['tag']\n",
        "  # # #appending rmsd values to df\n",
        "  # # df['RMSD_values_proteinx']= values\n",
        "\n",
        "\n",
        "  # # # x = df.loc[:, \"tag\"]\n",
        "  # # # y = df.loc[:, \"RMSD_values_proteinx\"]\n",
        "\n",
        "  # # # fig, ax = plt.subplots()\n",
        "  # # # ax.stem(x,y)\n",
        "  # # plt.ylim([0,5])\n",
        "  # # # ax.set(xlim=(0,0),xticks= np.arrange(1,5),\n",
        "  # # #       ylim=(0,10), yticks= np.arrange(1,10))\n",
        "  # # return fig\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 643
        },
        "id": "mEC2_ISr41Ze",
        "outputId": "5e26ddaa-6e4c-46a0-cffe-0426468bcb72"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Colab notebook detected. To show errors in colab notebook, set `debug=True` in `launch()`\n",
            "Running on public URL: https://17735.gradio.app\n",
            "\n",
            "This share link expires in 72 hours. For free permanent hosting, check out Spaces: https://huggingface.co/spaces\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<div><iframe src=\"https://17735.gradio.app\" width=\"900\" height=\"500\" allow=\"autoplay; camera; microphone;\" frameborder=\"0\" allowfullscreen></iframe></div>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(<gradio.routes.App at 0x7f7ecebfb810>,\n",
              " 'http://127.0.0.1:7860/',\n",
              " 'https://17735.gradio.app')"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ],
      "source": [
        "# Interface to determine number of proteins to compare where n is the number of proteins\n",
        "# 1 <= n <= 10\n",
        "\n",
        "files = {}\n",
        "\n",
        "def count_protein(quantity):\n",
        "    if (quantity == 1):\n",
        "        return gr.Textbox.update(lines=1, visible=True,placeholder=\"Enter protein sequence.\")\n",
        "    else:\n",
        "        return gr.Textbox.update(lines=quantity, visible=True,placeholder=\"Enter protein sequences separated by a semicolon\\nEnter a new sequence per line.\")\n",
        "\n",
        "def parse_choices(protein_num,seq_string,tag_choices):\n",
        "    filepaths_list = process_variants(process_str(seq_string),tag_choices)\n",
        "    native_list = input_dict(filepaths_list[0],True)\n",
        "    variant_list = input_dict(filepaths_list[1],False)\n",
        "    files = [native_list,variant_list]\n",
        "    plot = obtain_RMSD(native_list,variant_list)\n",
        "    return plot\n",
        "\n",
        "def process_str(seq_string):\n",
        "    seq_list = seq_string.split(\";\\n\")\n",
        "    return seq_list\n",
        "\n",
        "def process_variants(seq_list,tag_choices):\n",
        "    variant_list = {}\n",
        "    native_list = {}\n",
        "    for seq_index in range (0,len(seq_list)):\n",
        "        ori_seq = \"Protein\" + str(seq_index)\n",
        "        native_list[ori_seq] = seq_list[seq_index]\n",
        "        for tag in tag_choices:\n",
        "            variant_name = ori_seq + \"_\" + tag\n",
        "            variant_seq = seq_list[seq_index] + tag_sequences[tag]\n",
        "            variant_list[variant_name] = variant_seq\n",
        "    filepaths = [native_list,variant_list]\n",
        "    return filepaths\n",
        "        \n",
        "\n",
        "with gr.Blocks() as demo:\n",
        "    gr.Markdown(\n",
        "    \"\"\"\n",
        "    # Protein TAG Analyzer\n",
        "    Configure below:\n",
        "    \"\"\")\n",
        "    slider = gr.Slider(1,10,1,step=1,label=\"Number of Protein(s)\")\n",
        "    text = gr.Textbox(lines=1, interactive=True,placeholder=\"Enter protein sequence.\",label=\"Protein Sequence(s)\")\n",
        "    checkgroup = gr.CheckboxGroup([\"6X His\",\"CBP\",\"CBD + 3G\",\"mNEONGreen + 5G\"],label=\"Select TAGs\")\n",
        "    generate = gr.Button()\n",
        "    with gr.Row():\n",
        "            with gr.Column(scale=1):\n",
        "                rmsd_plot = gr.Plot()\n",
        "    \n",
        "    slider.change(fn=count_protein,inputs=slider,outputs=text)\n",
        "    generate.click(fn=parse_choices,inputs=[slider,text,checkgroup],outputs=rmsd_plot)\n",
        "\n",
        "demo.launch()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "files = [{'Protein0': '/content/PDBs/Protein0_9d99d_unrelaxed_rank_1_model_3.pdb','Protein1': '/content/PDBs/Protein1_f2bcc_unrelaxed_rank_1_model_4.pdb'}, {'Protein0_6XHis': '/content/PDBs/Protein0_6XHis_89f0d_unrelaxed_rank_1_model_1.pdb', 'Protein0_CBP': '/content/PDBs/Protein0_CBP_f392e_unrelaxed_rank_1_model_5.pdb',\n",
        "                                                                                     'Protein1_6XHis': '/content/PDBs/Protein1_6XHis_c8b1c_unrelaxed_rank_1_model_4.pdb', 'Protein1_CBP':'/content/PDBs/Protein1_CBP_1c6af_unrelaxed_rank_1_model_5.pdb'}]\n",
        "\n",
        "native_dict = files[0]\n",
        "variant_dict = files[1]\n",
        "\n",
        "rmsd_plot = obtain_RMSD(native_dict,variant_dict)\n",
        "rmsd_plot.legend()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 330
        },
        "id": "81Yzx26QmVHb",
        "outputId": "0655c0c0-6bd9-4cf5-f192-9292b65285c7"
      },
      "execution_count": 70,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[0 1]\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<matplotlib.legend.Legend at 0x7f7646e86cd0>"
            ]
          },
          "metadata": {},
          "execution_count": 70
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEWCAYAAABrDZDcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3de7xUZb3H8c9Xrl5QUzCVjYBieEPBtlivpLAsLU26WGlaKZpletSjXdTS1Drn2Dl2Uak4VqaWJaVmRF6ykrxU6EZRUUxRSTZeQhTECwc3/s4f69kwDDOzZ8teM8D6vl+v/dpr1rPWmt9as2Z+8zzPmmcpIjAzs+LaqNkBmJlZczkRmJkVnBOBmVnBORGYmRWcE4GZWcE5EZiZFZwTwQZG0nRJx63lNsZLal+L9SdLOnttYngDz/lhSfMlvSRpTCOfu9EkXS7pm82Oo16SdkivS69mx2KVFToRSJon6dV0kj6T3mCblZRfLikkTShb77tp/tHpcV9J35bUnrY1T9L3KjzPUkmLJf1V0uclrffHX9LRku4onRcRn4+IbzQ4lAuBkyJis4i4d2031hMJtajS+X5A5+OIeDK9LiuaGZdVt95/EPWAD0bEZsBoYAxwZln5I8CnOx9I6g18HHisZJkzgVZgLDAAGA/cU+F5BgBDgQuArwA/6bG9sKHAg29kRX9TtaJzIkgi4hngZrKEUOp3wH6S3pQeHwTcDzxTssw+wG8i4qnIzIuIK6s8z5KImAp8AviMpD0qLSdpK0k/lfSUpBckXZ/mv0nSNEkL0/xpklqq7Zekz0qak2ojD0naO80PSSNKlqva3CDpDEmPlWzjw2n+rsBk4O2pJrS40rZSDHMlPS9pqqTtS8oi1Y4eTbWl70tSKhsh6S+Slkh6TtKUCrH1k/QS0Au4T9JjnbGlb/WLJT0o6dCyff2hpBskvQzsX7bN/wDGAZPSfk1K8y9KzU8vSpopaVzJOhtLuiK9JnMkfbm0eU3S3pLuTcfw15KmlB2jQyTNKqkx7llSNkbSPWndKUD/Kq9Tv7T+HiXzBqXa6DaSBqbzZXF6LW6vVivt4nXZSdKfJS1Kr8tVkrZMZT8DdgB+l47dlyUNS9vrLekTktrKnuvfJU0t2YcLJT0p6VllzYwbV4mx6vkhaRdJt6T9/Iekj5eUbZ3Owxcl3SXpG0q12tJYS5ZfrXYoaWJ6jV+QdLOkofUct1Re7f24vaRrlb2vn5B0cqV9zk1EFPYPmAcckKZbgAeAi0rKLwe+CVwKnJDm/Qo4ArgDODrN+xrwJPAFYBSgas9TNv/Jzu1WKPs9MAV4E9AHeFeavzXwUWATstrHr4HrS9abDhyXpj8GLCBLVAJGAENTWQAjyvc1TY8H2kvKPgZsT/bF4RPAy8B2qexo4I6y2Eu39W7gOWBvoB9wCXBbybIBTAO2JPsAWQgclMp+CXw1PW9/YL8ar+XK/UnHay5wFtA3xbAUGFkS3xLgHZ3brrC9lcexZN5R6fj3Bk4n+zLQP5VdAPwlvV4tZF8W2lNZX+CfwCkpto8Ay0uO0RjgX8C+ZAntM+mc6Vey7r+ndQ8DXutct0LclwH/UfL4ROCmNP1fZIm7T/obR9m5WufrMgJ4b4pvEHAb8L1q5zswLG2vN9l5uxTYuaT8buDwNP1dYCqwFdn5/Tvgv6rEWPH8ADYF5gPHpOccQ3YO7pbKryZ7H28K7EH2HrmjPNYq76kJZOfWrmnbXwP+Wudxq/h+TPHPBM5Jr/eOwOPAgQ37LGzUE62Lf+mEfSmdmAH8CdiypPxyskSwH/C39OI+C2zM6omgV3rD3Qn8H/AU8Jlqb4yS+X8Hvlph/nbA68Cb6tiH0cALVU7am4FTqqxXdyKosO4sYEKaPpraieAnwH+XlG1G9kE2rCSO/UrKfwWckaavJEvCLXUch9JEMI7sQ3qjkvJfAueWxHdlF9tbeRxrLPMCsFeaXu2NCxzHqkTwzvQBoJLyO0qO0Q+Bb5Rt+x/Au9K6T5Wt+1eqJ4IDgMdKHt8JfDpNnw/8tvR17+J4VnxdKiz7IeDeauc7ZR+uwM+Bc9L0zmTvv03IPhxfBnYqWfftwBNVnrfi+UH2ZeX2snn/C3yd7L36GrBLSdl/Un8iuBE4tqRsI+AVVv+CVe18rvh+JPsC8GTZvDOBn3b1OvXUn5uG4EORtd2PB3YBBpYvEBF3kH3z+SowLSJeLStfERHfj4h3kCWL/wAuU9Z0Ustg4PkK84cAz0fEC+UFkjaR9L+S/inpRbJvY1uqcjv3EFbvy3hDJH26pNliMdm3qDWOUxXbk32jBSAiXgIWke17p9JmtlfIkgXAl8k+HO5S1rwzsRvPOT8iXi+Z98+y55xf57ZWkvTFVK1fko7DFqw6DtuXbbN0entgQaR3eIXyocDpncc3bXtIWq/Suv+kuluBTSTtK2kY2ReF36Sy/yH7NvsHSY9LOqOLXa74ukh6s6SrJS1I5+DPqf98APgFWa0a4JNkNdpXyN5jmwAzS47DTWl+JdXOj6HAvmXH80hg27St3qx+/Gsdz3JDgYtKtvt8iqGe87na+3EosH1ZvGcBb+5GXGvFiSCJiL+QfVO8sMoiPydrDqjY9l+ynVcj4vtk3xZ3q7acpH3ITp47KhTPB7bqbHctczowEtg3IjYn+8YI2clYaTs7VQnhFbI3Xadtq8Q5FPgRcBKwdURsCcwueb6otF6Jp8hO9M7tbUrWvLKgi/WIiGci4rMRsT3wOeAHKunX6OI5h5S1f+9Q9pxdxb1aubL+gC+TXSjwpnQclrDqODxN1iTUaUjJ9NPA4NK24rLy+WTNOVuW/G0SEb+ssu4OVYPOrszpbL48guyLy9JUtjQiTo+IHYFDgdMkvaf2YajoP8mOz6h0Dh7F6udfV8f2FmCQpNEpxl+k+c8BrwK7lxyHLSK7mKPSvlY7P+YDfyk7nptFxAlkTTUdrH78S4/ny+l/tffGfOBzZdveOCL+2sU+d65b6f04n6zWU7rNARHxgTq22SOcCFb3PeC9kvaqUHYxWbvobeUFkk5Vdu39xqlD7DNk7ZtrXMYoaXNJh5C1U/48Ih4oXyYiniargv5AWedwH0mdH/gDyN4siyVtRVbdrebHwBclvVWZESUdW7OAT0rqJekgsmaISjYle2MvTPEfQ1Yj6PQs0CKpb5X1fwkcI2m0pH5kHyIzImJejbhJz/UxreoIfyHF8XqNVTrNIEt0X07HbjzwQbJjXq9nydpqOw0g+wBZCPSWdA6weUn5r4Az0+s1mCxxdvobsAI4KZ0fE8iuMOv0I+Dz6Vu8JG0q6WBJA9K6HcDJaV8+UrZuJb8gax45klUfsp0d0iNSUlmSYqrneJYbQNakuiTt65fKysuP3Woi4jWyvq3/IesLuCXNf53sWHxX0jYp5sGSDqy0nRrnxzTgLZI+lY5ZH0n7SNo1JcrrgHNT7Xo3sj6ZztgWkn1hOCq9Nyay+of3ZLLXefcUwxaSPlb7cK1U7f14F7BU0lfSZ0gvSXukL4sN4URQIp0EV5J12pSXPR8Rfyqrond6Bfg2WZXwObL+go9GxOMly/xO0lKy7P9V4DtknVnVfIqsLfNhso7EU9P875H1UTxH1sdwU439+TVZM9UvyNphryd740HWcflBoLPafH2VbTyU9u1vZG/wUWTtzp3+THbZ5jOSnquw/h+Bs4Fryb7d7gQcXmO/S+0DzFB2VdBUsvbVx7tYh4hYnvbt/WTH6Qdk7eQP1/m8ABcBhym7MuRisvbdm8guJ/4nsIzVmxfOB9qBJ4A/AteQ9Rd1xvMR4Fiy430U2YdVZ3kb8FlgEtkH2lyyvpfSdY8ma4b4BNkHWa39n0H2zXZ7si8UnXZOsb1E9nr+ICJu7cYx6XQeWef/ErKLGsrj+S/ga6mZ44tVtvELsv6MX0dER8n8r5Dt/99Ts9MfyWrAlVQ8P1IN6H1k59lTZO/Lb5F1bkOWpDdL8y8Hflq23c+SJbdFwO5kfTIARMRv0rauTvHNJjvPulTt/ZiS0yFkzXhPkJ2zPyZremwIVf5cM7O1IekEsithKta0JM0AJkdE+YeQNZiyH4YeFxH7NTuWZnGNwKwHSNpO0jskbSRpJFlfzm9Kyt8laduSpsM9qVGbM2uk3l0vYmZ16Et2ieJwsuafq8mapDqNZNW1648Dh6W+ILOmc9OQmVnBuWnIzKzg1rumoYEDB8awYcOaHYaZ2Xpl5syZz0VExR/nrXeJYNiwYbS1tXW9oJmZrSSp6i+o3TRkZlZwuSUCSf2VDfF6XxoH5LwKy/RTNhzvXEkzlI2NYmZmDZRnjeD/gHdHxF5kv5g7SNLbypY5lmzkzBFkw89+K8d4zMysgtz6CNJQDC+lh53jn5dfqzoBODdNX0N2IxBVGcahqtdee4329naWLVu2FhEXS//+/WlpaaFPnz7NDsXMmizXzmJlQyPPJLsBw/fTGCilBpPGa4mIDklLyEamXGPMmlra29sZMGAAw4YNY/VBGq2SiGDRokW0t7czfPjwZodjZk2Wa2dxGqd/NNnwvGNV5baMXZF0vKQ2SW0LFy5co3zZsmVsvfXWTgJ1ksTWW2/tGpSZAQ26aigiFpPdMOOgsqIFpHHBld0jdAuyEf/K1780IlojonXQoMr3qHAS6B4fLzPrlOdVQ4O06obWG5ON5V8+DPBUVo0Ffhjw5+72D5iZ2drJs49gO+CK1E+wEfCriJgm6XygLSKmkt3P9meS5pKNtV7vOPU1DTvj9z2xmZXmXXBwl8v06tWLUaNG0dHRwa677soVV1zBJpts0uV6ALNmzeKpp57iAx+ofUOitrY2rrzySi6++OKay910002ccsoprFixguOOO44zzujqjoRmVmR5XjV0PzCmwvxzSqaXAfXe3WedtvHGGzNr1iwAjjzySCZPnsxpp522sryjo4PevSsf7lmzZtHW1tZlImhtbaW1tbXmMitWrODEE0/klltuoaWlhX322YdDDz2U3XaretdMs3XXuQ27N8v64dwluWzWvyzOwbhx45g7dy7Tp09n3LhxKz+Ily1bxjHHHMOoUaMYM2YMt956K8uXL+ecc85hypQpjB49milTpvDyyy8zceJExo4dy5gxY/jtb38LwPTp0znkkEMAOPfcc5k4cSLjx49nxx13XFlLuOuuuxgxYgQ77rgjffv25fDDD1+5vplZJevdWEPruo6ODm688UYOOijrF7/nnnuYPXs2w4cP59vf/jaSeOCBB3j44Yd53/vexyOPPML5559PW1sbkyZNAuCss87i3e9+N5dddhmLFy9m7NixHHDAAWs818MPP8ytt97K0qVLGTlyJCeccAILFixgyJBV9+VuaWlhxozyq3bNzFZxjaCHvPrqq4wePZrW1lZ22GEHjj32WADGjh278lr9O+64g6OOOgqAXXbZhaFDh/LII4+ssa0//OEPXHDBBYwePZrx48ezbNkynnzyyTWWO/jgg+nXrx8DBw5km2224dlnn81xD81sQ+UaQQ8p7SMotemmm3Z7WxHBtddey8iRq9+zu/yDvl+/fiune/XqRUdHB4MHD2b+/FX3VG9vb2fw4MHdjsHMisM1ggYaN24cV111FQCPPPIITz75JCNHjmTAgAEsXbp05XIHHnggl1xyCZ1X0t577711P8c+++zDo48+yhNPPMHy5cu5+uqrOfTQQ3t2R8xsg7JB1gjqudyzGb7whS9wwgknMGrUKHr37s3ll19Ov3792H///Vc2BZ155pmcffbZnHrqqey55568/vrrDB8+nGnTptX1HL1792bSpEkceOCBrFixgokTJ7L77rvnvGfWk3r68uf12bz+zY6gGNa7exa3trZG+Y1p5syZw6677tqkiNZfPm7rJieCVeb1/2SzQ1i3rMXlo5JmRkTF68/dNGRmVnBOBGZmBedEYGZWcE4EZmYF50RgZlZwTgRmZgW3Qf6OoMdHLKzjkq11aRjqiRMnMm3aNLbZZhtmz55dVwxmVlyuEfSQziEmZs+eTd++fZk8efJq5R0dHVXXnTVrFjfccEOXz9Ha2tplEgA4+uijuemmm7oO2swMJ4JcNHMYaoB3vvOdbLXVVk3ZdzNb/2yYTUNN1OxhqPv06dPQ/TWz9Z9rBD3Ew1Cb2frKNYIesq4MQ21m1l2uETRQI4ahNjPrrg2zRpDTDZ7XViOGoQY44ogjmD59Os899xwtLS2cd955K5uqzMzKeRjqAvNxWzd5GOpVPAx1GQ9DbWZmeXAiMDMruA0mEaxvTVzN5uNlZp02iETQv39/Fi1a5A+3OkUEixYton9/3xDWzDaQq4ZaWlpob29n4cKFzQ5lvdG/f39aWlqaHYaZrQNySwSShgBXAm8GArg0Ii4qW2Y88FvgiTTruog4v7vP1adPn5W/3jUzs+7Js0bQAZweEfdIGgDMlHRLRDxUttztEXFIjnGYmVkNufURRMTTEXFPml4KzAEG5/V8Zmb2xjSks1jSMGAMMKNC8dsl3SfpRkm7V1n/eEltktrcD2Bm1rNyTwSSNgOuBU6NiBfLiu8BhkbEXsAlwPWVthERl0ZEa0S0Dho0KN+AzcwKJtdEIKkPWRK4KiKuKy+PiBcj4qU0fQPQR9LAPGMyM7PV5ZYIJAn4CTAnIr5TZZlt03JIGpviWZRXTGZmtqY8rxp6B/Ap4AFJnQP1nwXsABARk4HDgBMkdQCvAoeHfxVmZtZQuSWCiLgDUBfLTAIm5RWDmZl1bYMYYsLMzN44JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4LLLRFIGiLpVkkPSXpQ0ikVlpGkiyXNlXS/pL3zisfMzCrrneO2O4DTI+IeSQOAmZJuiYiHSpZ5P7Bz+tsX+GH6b2ZmDZJbjSAino6Ie9L0UmAOMLhssQnAlZH5O7ClpO3yisnMzNbUkD4CScOAMcCMsqLBwPySx+2smSyQdLykNkltCxcuzCtMM7NCyj0RSNoMuBY4NSJefCPbiIhLI6I1IloHDRrUswGamRVcl4lA0sdSGz+Svibpuno7dSX1IUsCV0XEdRUWWQAMKXnckuaZmVmD1FMjODsilkraDzgA+AlZp25NkpSWnRMR36my2FTg0+nqobcBSyLi6TpjNzOzHlDPVUMr0v+DgUsj4veSvlnHeu8APgU8IGlWmncWsANAREwGbgA+AMwFXgGO6UbsZmbWA+pJBAsk/S/wXuBbkvpRR00iIu4A1MUyAZxYT6BmZpaPepqGPg7cDBwYEYuBrYAv5RqVmZk1TD3f7F8B/gXsl2Z1AI/mGZSZmTVOPVcNfR34CnBmmtUH+HmeQZmZWePU0zT0YeBQ4GWAiHgKGJBnUGZm1jj1JILlqVM3ACRtmm9IZmbWSPUkgl+lq4a2lPRZ4I/Aj/INy8zMGqXLy0cj4kJJ7wVeBEYC50TELblHZmZmDVHXMNTpg98f/mZmG6AuE4GkpaT+AaAv2VVDL0fE5nkGZmZmjVFP09DKK4TS+EETgLflGZSZmTVOt4ahTjeQuR44MKd4zMysweppGvpIycONgFZgWW4RmZlZQ9XTWfzBkukOYB5Z85CZmW0A6ukj8NDQZmYbsKqJQNIlrLpaaA0RcXIuEZmZWUPVqhG0NSwKMzNrmqqJICKuaGQgZmbWHPVcNTSIbBjq3YD+nfMj4t05xmVmZg1Sz+8IrgLmAMOB88iuGro7x5jMzKyB6kkEW0fET4DXIuIvETERcG3AzGwDUc/vCF5L/5+WdDDwFNl9i83MbANQTyL4pqQtgNOBS4DNgX/PNSozM2uYehLBjIhYAiwB9s85HjMza7B6EsGdkuYBU4DrIuKFfEPKz7Azft/sENYp8y44uNkhmNk6oMvO4oh4C/A1YHdgpqRpko7KPTIzM2uIuoahjoi7IuI0YCzwPOAfm5mZbSC6TASSNpf0GUk3An8FniZLCF2td5mkf0maXaV8vKQlkmalv3O6Hb2Zma21evoI7gOuB86PiL91Y9uXA5OAK2ssc3tEHNKNbZqZWQ+rJxHsGBFVRyGtJiJukzSs2xGZmVlD1dNZ3O0k0A1vl3SfpBsl7V5tIUnHS2qT1LZw4cIcwzEzK55u3bO4h90DDI2Ivch+qHZ9tQUj4tKIaI2I1kGDBjUsQDOzImhaIoiIFyPipTR9A9BH0sBmxWNmVlQ1E4Gk/SVdJ+nB9HeNpPE98cSStpWkND02xbKoJ7ZtZmb1q3WryoPJrvo5n2z4aQF7A5dJOil9i69K0i+B8cBASe3A14E+ABExGTgMOEFSB/AqcHjO/RFmZlZBrauGvgR8KCLuK5k3S1IbWZt+zUQQEUd0UT6JLNGYmVkT1Woa2rYsCQAQEfcDb84vJDMza6RaieDlN1hmZmbrkVpNQztJmlphvoAdc4rHzMwarFYimFCj7MKeDsTMzJqjaiKIiL+UPpbUB9gDWBAR/8o7MDMza4yqfQSSJncO+5BuVXkf2QBy90qqeUWQmZmtP2p1Fo+LiAfT9DHAIxExCngr8OXcIzMzs4aolQiWl0y/lzQWUEQ8k2tEZmbWULUSwWJJh0gaA7wDuAlAUm9g40YEZ2Zm+at11dDngIuBbYFTS2oC7wF8F3gzsw1ErauGHgEOqjD/ZuDmPIMyM7PGqTXo3MW1VoyIk3s+HDMza7RaTUOfB2YDvwKeIvtFsZmZbWBqJYLtgI8BnwA6gCnANRGxuBGBmZlZY1S9aigiFkXE5IjYn+x3BFsCD0n6VMOiMzOz3NWqEQAgaW/gCLLfEtwIzMw7KDMza5xancXnAwcDc4CrgTMjoqNRgZmZWWPUqhF8DXgC2Cv9/WfnLYaBiIg98w/PzMzyVisRDG9YFGZm1jS1flD2z0rzJW1E1mdQsdzMzNYvtYah3lzSmZImSXqfMv8GPA58vHEhmplZnmo1Df0MeAH4G3AccBZZ/8CHImJWA2IzM7MGqJUIdkz3H0DSj4GngR0iYllDIjMzs4aoNQz1a50TEbECaHcSMDPb8NSqEewl6cU0LWDj9Ljz8tHNc4/OzMxyV+uqoV6NDMTMzJqjVtPQWpF0maR/SZpdpVySLpY0V9L9aSgLMzNrsNwSAXA5FW5sU+L9wM7p73jghznGYmZmVeSWCCLiNuD5GotMAK6MzN+BLSVtl1c8ZmZWWZ41gq4MBuaXPG5P89Yg6XhJbZLaFi5c2JDgzMyKopmJoG4RcWlEtEZE66BBg5odjpnZBqWZiWABMKTkcUuaZ2ZmDdTMRDAV+HS6euhtwJKIeLqJ8ZiZFVKXdyh7oyT9EhgPDJTUDnwd6AMQEZOBG4APAHOBV8huh2lmZg2WWyKIiCO6KA/gxLye38zM6rNedBabmVl+nAjMzArOicDMrOCcCMzMCs6JwMys4JwIzMwKzonAzKzgnAjMzArOicDMrOCcCMzMCs6JwMys4JwIzMwKzonAzKzgnAjMzArOicDMrOByux+BrQfO3aLZEaxbzl3S7AjMmsI1AjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMyu4XBOBpIMk/UPSXElnVCg/WtJCSbPS33F5xmNmZmvKbawhSb2A7wPvBdqBuyVNjYiHyhadEhEn5RWHmZnVlmeNYCwwNyIej4jlwNXAhByfz8zM3oA8E8FgYH7J4/Y0r9xHJd0v6RpJQ3KMx8zMKmh2Z/HvgGERsSdwC3BFpYUkHS+pTVLbwoULGxqgmdmGLs9EsAAo/YbfkuatFBGLIuL/0sMfA2+ttKGIuDQiWiOiddCgQbkEa2ZWVHkmgruBnSUNl9QXOByYWrqApO1KHh4KzMkxHjMzqyC3q4YiokPSScDNQC/gsoh4UNL5QFtETAVOlnQo0AE8DxydVzxmZlZZrreqjIgbgBvK5p1TMn0mcGaeMZiZWW3N7iw2M7MmcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4LLNRFIOkjSPyTNlXRGhfJ+kqak8hmShuUZj5mZrSm3RCCpF/B94P3AbsARknYrW+xY4IWIGAF8F/hWXvGYmVlledYIxgJzI+LxiFgOXA1MKFtmAnBFmr4GeI8k5RiTmZmV6Z3jtgcD80setwP7VlsmIjokLQG2Bp4rXUjS8cDx6eFLkv6RS8QFsw5l3IGUveZNcd46dEQM8Dm6hrU7R4dWK8gzEfSYiLgUuLTZcVg+JLVFRGuz4zCrZkM/R/NsGloADCl53JLmVVxGUm9gC2BRjjGZmVmZPBPB3cDOkoZL6gscDkwtW2Yq8Jk0fRjw54iIHGMyM7MyuTUNpTb/k4CbgV7AZRHxoKTzgbaImAr8BPiZpLnA82TJworHzX62rtugz1H5C7iZWbH5l8VmZgXnRGBmVnBOBNajJG0p6RpJD0uaI+ntkqZK+nTJMj+S9KU0PV1Sa0nZMEmz03SrpIsbvxdWNJK2lXS1pMckzZR0g6S3SHpV0ixJ90n6q6SRafnxkpaksjmSvt7sfVgb68XvCGy9chFwU0Qclq4W2wQ4GbhV0lSy4Ub2BU7oakMR0Qa05RmsWRrN4DfAFRFxeJq3F/Bm4LGIGJ3mfQ44i1VXOt4eEYdI2hSYJel3EXFP4/dg7blGYD1G0hbAO8muBiMilkfE4oiYR3bVxX8DPwROioiOOrY3XtK0NP2u9O1rlqR7JQ3IbUesaPYHXouIyZ0zIuI+Vh8ZAWBz4IXylSPiZWAmMCLPIPPkGoH1pOHAQuCn6RvVTOCU9Ea5EHiM7FvUbWXrXSXp1TTdF3i9wra/CJwYEXdK2gxYlsseWBHtQWhLsaYAAAK/SURBVHauVrKTpFnAALLabfkwOUjaGngb8I3cIsyZawTWk3oDewM/jIgxwMtA5/Dje5Kdb7tIKj/vjoyI0akK/oEq274T+I6kk4Et66lRmPWAx9K5uRNwKqv/nmCcpHuBPwAXRMSDTYmwBzgRWE9qB9ojYkZ6fA2wd/rg/wFwFPAodfQPlIuIC4DjgI2BOyXt0jMhm/Eg8NY6lptK1vTZ6faIGBMRby1tVlofORFYj4mIZ4D5nVdWAO8BHgI+BzwaEdOB04CvSBrUnW1L2ikiHoiIb5ENX+JEYD3lz0C/NMoxAJL2ZPWx0gD2I2ve3OC4j8B62r+Rtfn3BR4nSwIzyNpQiYinJH2PrOP4mG5s91RJ+5P1HzwI3NijUVthRURI+jDwPUlfIet/mkfWFNTZRyBgOVmtdIPjISbMzArOTUNmZgXnRGBmVnBOBGZmBedEYGZWcE4EZmYF58tHzWpIwwf8KT3cFlhBNowGwNiIWN6UwMx6kC8fNauTpHOBlyLiwmbHYtaT3DRk1k2SPivp7jRG/bWSNknzd5L0d0kPSPqmpJfS/O0k3ZZGTp0taVxz98BsdU4EZt13XUTsExF7AXOAY9P8i4CLImIU2bhLnT4J3JwG1dsLmNXQaM264ERg1n17SLpd0gPAkcDuaf7bgV+n6V+ULH83cExqWhoVEUsbFqlZHZwIzLrvcrKb64wCzgP611o43X/hncAC4PLS23aarQucCMy6bwDwtKQ+ZDWCTn8HPpqmD++cKWko8GxE/Aj4Mdk9G8zWGU4EZt13NtmIqncCD5fMPxU4TdL9ZLctXJLmjwfuSzcx+QRZX4LZOsOXj5r1kHT10KtpWOPDgSMiYkKz4zLrin9QZtZz3gpMkiRgMTCxyfGY1cU1AjOzgnMfgZlZwTkRmJkVnBOBmVnBORGYmRWcE4GZWcH9PzAx8rPmvAssAAAAAElFTkSuQmCC\n"
          },
          "metadata": {
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "gR8CUxXo8RAW"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "collapsed_sections": [],
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.13"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}